// @mistralai/mistralai/esm/models/errors/responsevalidationerror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 88ff98a41be9
 */
import * as z from "zod/v4/core";
import { MistralError } from "./mistralerror.js";
import { formatZodError } from "./sdkvalidationerror.js";
export class ResponseValidationError extends MistralError {
    /**
     * The raw value that failed validation.
     */
    rawValue;
    /**
     * The raw message that failed validation.
     */
    rawMessage;
    constructor(message, extra) {
        super(message, extra);
        this.name = "ResponseValidationError";
        this.cause = extra.cause;
        this.rawValue = extra.rawValue;
        this.rawMessage = extra.rawMessage;
    }
    /**
     * Return a pretty-formatted error message if the underlying validation error
     * is a ZodError or some other recognized error type, otherwise return the
     * default error message.
     */
    pretty() {
        if (this.cause instanceof z.$ZodError) {
            return `${this.rawMessage}\n${formatZodError(this.cause)}`;
        }
        else {
            return this.toString();
        }
    }
}
