// @mistralai/mistralai/src/models/components/turbinemeta.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7becc1559bde
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import { ServerLocale, ServerLocale$inboundSchema } from "./serverlocale.js";

export type TurbineMeta = {
  systemPromptName?: string | null | undefined;
  locale?: ServerLocale | null | undefined;
};

/** @internal */
export const TurbineMeta$inboundSchema: z.ZodType<TurbineMeta, unknown> = z
  .object({
    system_prompt_name: z.nullable(z.string()).optional(),
    locale: z.nullable(ServerLocale$inboundSchema).optional(),
  }).transform((v) => {
    return remap$(v, {
      "system_prompt_name": "systemPromptName",
    });
  });

export function turbineMetaFromJSON(
  jsonString: string,
): SafeParseResult<TurbineMeta, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => TurbineMeta$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'TurbineMeta' from JSON`,
  );
}
