// @mistralai/mistralai/packages/mistralai-azure/src/models/components/customconnector.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6687cfaf9029
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import {
  APIKeyAuth,
  APIKeyAuth$Outbound,
  APIKeyAuth$outboundSchema,
} from "./apikeyauth.js";
import {
  OAuth2TokenAuth,
  OAuth2TokenAuth$Outbound,
  OAuth2TokenAuth$outboundSchema,
} from "./oauth2tokenauth.js";
import {
  ToolConfiguration,
  ToolConfiguration$Outbound,
  ToolConfiguration$outboundSchema,
} from "./toolconfiguration.js";

export type Authorization = APIKeyAuth | OAuth2TokenAuth;

export type CustomConnector = {
  type: "connector";
  connectorId: string;
  authorization?: APIKeyAuth | OAuth2TokenAuth | null | undefined;
  toolConfiguration?: ToolConfiguration | null | undefined;
};

/** @internal */
export type Authorization$Outbound =
  | APIKeyAuth$Outbound
  | OAuth2TokenAuth$Outbound;

/** @internal */
export const Authorization$outboundSchema: z.ZodType<
  Authorization$Outbound,
  Authorization
> = z.union([APIKeyAuth$outboundSchema, OAuth2TokenAuth$outboundSchema]);

export function authorizationToJSON(authorization: Authorization): string {
  return JSON.stringify(Authorization$outboundSchema.parse(authorization));
}

/** @internal */
export type CustomConnector$Outbound = {
  type: "connector";
  connector_id: string;
  authorization?:
    | APIKeyAuth$Outbound
    | OAuth2TokenAuth$Outbound
    | null
    | undefined;
  tool_configuration?: ToolConfiguration$Outbound | null | undefined;
};

/** @internal */
export const CustomConnector$outboundSchema: z.ZodType<
  CustomConnector$Outbound,
  CustomConnector
> = z.object({
  type: z.literal("connector"),
  connectorId: z.string(),
  authorization: z.nullable(
    z.union([APIKeyAuth$outboundSchema, OAuth2TokenAuth$outboundSchema]),
  ).optional(),
  toolConfiguration: z.nullable(ToolConfiguration$outboundSchema).optional(),
}).transform((v) => {
  return remap$(v, {
    connectorId: "connector_id",
    toolConfiguration: "tool_configuration",
  });
});

export function customConnectorToJSON(
  customConnector: CustomConnector,
): string {
  return JSON.stringify(CustomConnector$outboundSchema.parse(customConnector));
}
