// @mistralai/mistralai/packages/mistralai-gcp/src/models/components/codeinterpretertool.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d988a9992fe2
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import {
  ToolConfiguration,
  ToolConfiguration$Outbound,
  ToolConfiguration$outboundSchema,
} from "./toolconfiguration.js";

export type CodeInterpreterTool = {
  toolConfiguration?: ToolConfiguration | null | undefined;
  type: "code_interpreter";
};

/** @internal */
export type CodeInterpreterTool$Outbound = {
  tool_configuration?: ToolConfiguration$Outbound | null | undefined;
  type: "code_interpreter";
};

/** @internal */
export const CodeInterpreterTool$outboundSchema: z.ZodType<
  CodeInterpreterTool$Outbound,
  CodeInterpreterTool
> = z.object({
  toolConfiguration: z.nullable(ToolConfiguration$outboundSchema).optional(),
  type: z.literal("code_interpreter"),
}).transform((v) => {
  return remap$(v, {
    toolConfiguration: "tool_configuration",
  });
});

export function codeInterpreterToolToJSON(
  codeInterpreterTool: CodeInterpreterTool,
): string {
  return JSON.stringify(
    CodeInterpreterTool$outboundSchema.parse(codeInterpreterTool),
  );
}
