// @mistralai/mistralai/esm/models/operations/agentsapiv1agentslist.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 417647f39e0b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const AgentsApiV1AgentsListRequest$outboundSchema = z.object({
    page: z.int().default(0),
    pageSize: z.int().default(20),
    deploymentChat: z.nullable(z.boolean()).optional(),
    sources: z.nullable(z.array(components.RequestSource$outboundSchema))
        .optional(),
    name: z.nullable(z.string()).optional(),
    search: z.nullable(z.string()).optional(),
    id: z.nullable(z.string()).optional(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        deploymentChat: "deployment_chat",
    });
});
export function agentsApiV1AgentsListRequestToJSON(agentsApiV1AgentsListRequest) {
    return JSON.stringify(AgentsApiV1AgentsListRequest$outboundSchema.parse(agentsApiV1AgentsListRequest));
}
