// @mistralai/mistralai/esm/models/operations/agentsapiv1conversationslist.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9a7100ad3737
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const AgentsApiV1ConversationsListRequest$outboundSchema = z.object({
    page: z.int().default(0),
    pageSize: z.int().default(100),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
    });
});
export function agentsApiV1ConversationsListRequestToJSON(agentsApiV1ConversationsListRequest) {
    return JSON.stringify(AgentsApiV1ConversationsListRequest$outboundSchema.parse(agentsApiV1ConversationsListRequest));
}
/** @internal */
export const AgentsApiV1ConversationsListResponse$inboundSchema = smartUnion([
    components.ModelConversation$inboundSchema,
    components.AgentConversation$inboundSchema,
]);
export function agentsApiV1ConversationsListResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AgentsApiV1ConversationsListResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AgentsApiV1ConversationsListResponse' from JSON`);
}
