// @mistralai/mistralai/esm/models/operations/connectorlisttoolsv1.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d8f478d65695
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const ConnectorListToolsV1Request$outboundSchema = z.object({
    page: z.int().default(1),
    pageSize: z.int().default(100),
    refresh: z.boolean().default(false),
    pretty: z.boolean().default(false),
    credentialsName: z.nullable(z.string()).optional(),
    connectorIdOrName: z.string(),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        credentialsName: "credentials_name",
        connectorIdOrName: "connector_id_or_name",
    });
});
export function connectorListToolsV1RequestToJSON(connectorListToolsV1Request) {
    return JSON.stringify(ConnectorListToolsV1Request$outboundSchema.parse(connectorListToolsV1Request));
}
/** @internal */
export const ResponseConnectorListToolsV1$inboundSchema = smartUnion([
    z.array(components.ConnectorTool$inboundSchema),
    z.array(components.MCPTool$inboundSchema),
    z.array(z.record(z.string(), z.any())),
]);
export function responseConnectorListToolsV1FromJSON(jsonString) {
    return safeParse(jsonString, (x) => ResponseConnectorListToolsV1$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ResponseConnectorListToolsV1' from JSON`);
}
