// @mistralai/mistralai/esm/models/operations/filesapirouteslistfiles.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c87977aa89c7
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const FilesApiRoutesListFilesRequest$outboundSchema = z.object({
    page: z.int().default(0),
    pageSize: z.int().default(100),
    includeTotal: z.boolean().default(true),
    sampleType: z.nullable(z.array(components.SampleType$outboundSchema))
        .optional(),
    source: z.nullable(z.array(components.Source$outboundSchema)).optional(),
    search: z.nullable(z.string()).optional(),
    purpose: z.nullable(components.FilePurpose$outboundSchema).optional(),
    mimetypes: z.nullable(z.array(z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        includeTotal: "include_total",
        sampleType: "sample_type",
    });
});
export function filesApiRoutesListFilesRequestToJSON(filesApiRoutesListFilesRequest) {
    return JSON.stringify(FilesApiRoutesListFilesRequest$outboundSchema.parse(filesApiRoutesListFilesRequest));
}
