// @mistralai/mistralai/esm/models/operations/getjudgesv1observabilityjudgesget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9e30c8d35c7a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
/** @internal */
export const GetJudgesV1ObservabilityJudgesGetRequest$outboundSchema = z.object({
    typeFilter: z.nullable(z.array(components.JudgeOutputType$outboundSchema))
        .optional(),
    modelFilter: z.nullable(z.array(z.string())).optional(),
    pageSize: z.int().default(50),
    page: z.int().default(1),
    q: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        typeFilter: "type_filter",
        modelFilter: "model_filter",
        pageSize: "page_size",
    });
});
export function getJudgesV1ObservabilityJudgesGetRequestToJSON(getJudgesV1ObservabilityJudgesGetRequest) {
    return JSON.stringify(GetJudgesV1ObservabilityJudgesGetRequest$outboundSchema.parse(getJudgesV1ObservabilityJudgesGetRequest));
}
