// @mistralai/mistralai/esm/models/operations/getschedulesv1workflowsschedulesget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9b8d9812bc5d
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/**
 * Filter by schedule status: 'active' or 'paused'
 */
export const GetSchedulesV1WorkflowsSchedulesGetStatus = {
    Active: "active",
    Paused: "paused",
};
/** @internal */
export const GetSchedulesV1WorkflowsSchedulesGetStatus$outboundSchema = z.enum(GetSchedulesV1WorkflowsSchedulesGetStatus);
/** @internal */
export const GetSchedulesV1WorkflowsSchedulesGetRequest$outboundSchema = z.object({
    workflowName: z.nullable(z.string()).optional(),
    userId: z.nullable(z.string()).optional(),
    status: z.nullable(GetSchedulesV1WorkflowsSchedulesGetStatus$outboundSchema)
        .optional(),
    pageSize: z.nullable(z.int()).optional(),
    nextPageToken: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        workflowName: "workflow_name",
        userId: "user_id",
        pageSize: "page_size",
        nextPageToken: "next_page_token",
    });
});
export function getSchedulesV1WorkflowsSchedulesGetRequestToJSON(getSchedulesV1WorkflowsSchedulesGetRequest) {
    return JSON.stringify(GetSchedulesV1WorkflowsSchedulesGetRequest$outboundSchema.parse(getSchedulesV1WorkflowsSchedulesGetRequest));
}
/** @internal */
export const GetSchedulesV1WorkflowsSchedulesGetResponse$inboundSchema = z.object({
    Result: components.WorkflowScheduleListResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function getSchedulesV1WorkflowsSchedulesGetResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetSchedulesV1WorkflowsSchedulesGetResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetSchedulesV1WorkflowsSchedulesGetResponse' from JSON`);
}
