// @mistralai/mistralai/esm/models/operations/getworkflowsv1workflowsget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e8470b4d34a8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as components from "../components/index.js";
/** @internal */
export const GetWorkflowsV1WorkflowsGetRequest$outboundSchema = z.object({
    activeOnly: z.boolean().default(false),
    includeShared: z.boolean().default(true),
    availableInChatAssistant: z.nullable(z.boolean()).optional(),
    archived: z.nullable(z.boolean()).optional(),
    cursor: z.nullable(z.string()).optional(),
    limit: z.int().default(50),
}).transform((v) => {
    return remap$(v, {
        activeOnly: "active_only",
        includeShared: "include_shared",
        availableInChatAssistant: "available_in_chat_assistant",
    });
});
export function getWorkflowsV1WorkflowsGetRequestToJSON(getWorkflowsV1WorkflowsGetRequest) {
    return JSON.stringify(GetWorkflowsV1WorkflowsGetRequest$outboundSchema.parse(getWorkflowsV1WorkflowsGetRequest));
}
/** @internal */
export const GetWorkflowsV1WorkflowsGetResponse$inboundSchema = z.object({
    Result: components.WorkflowListResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function getWorkflowsV1WorkflowsGetResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetWorkflowsV1WorkflowsGetResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetWorkflowsV1WorkflowsGetResponse' from JSON`);
}
