// @mistralai/mistralai/esm/models/operations/getworkflowregistrationsv1workflowsregistrationsget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 204467da11b7
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/** @internal */
export const GetWorkflowRegistrationsV1WorkflowsRegistrationsGetRequest$outboundSchema = z.object({
    workflowId: z.nullable(z.string()).optional(),
    taskQueue: z.nullable(z.string()).optional(),
    activeOnly: z.boolean().default(false),
    includeShared: z.boolean().default(true),
    workflowSearch: z.nullable(z.string()).optional(),
    archived: z.nullable(z.boolean()).optional(),
    withWorkflow: z.boolean().default(false),
    availableInChatAssistant: z.nullable(z.boolean()).optional(),
    limit: z.int().default(50),
    cursor: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        workflowId: "workflow_id",
        taskQueue: "task_queue",
        activeOnly: "active_only",
        includeShared: "include_shared",
        workflowSearch: "workflow_search",
        withWorkflow: "with_workflow",
        availableInChatAssistant: "available_in_chat_assistant",
    });
});
export function getWorkflowRegistrationsV1WorkflowsRegistrationsGetRequestToJSON(getWorkflowRegistrationsV1WorkflowsRegistrationsGetRequest) {
    return JSON.stringify(GetWorkflowRegistrationsV1WorkflowsRegistrationsGetRequest$outboundSchema
        .parse(getWorkflowRegistrationsV1WorkflowsRegistrationsGetRequest));
}
