// @mistralai/mistralai/esm/models/operations/getworkfloweventsv1workflowseventslistget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c0ec549bcb71
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/** @internal */
export const GetWorkflowEventsV1WorkflowsEventsListGetRequest$outboundSchema = z.object({
    rootWorkflowExecId: z.nullable(z.string()).optional(),
    workflowExecId: z.nullable(z.string()).optional(),
    workflowRunId: z.nullable(z.string()).optional(),
    limit: z.int().default(100),
    cursor: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        rootWorkflowExecId: "root_workflow_exec_id",
        workflowExecId: "workflow_exec_id",
        workflowRunId: "workflow_run_id",
    });
});
export function getWorkflowEventsV1WorkflowsEventsListGetRequestToJSON(getWorkflowEventsV1WorkflowsEventsListGetRequest) {
    return JSON.stringify(GetWorkflowEventsV1WorkflowsEventsListGetRequest$outboundSchema.parse(getWorkflowEventsV1WorkflowsEventsListGetRequest));
}
