// @mistralai/mistralai/esm/models/operations/jobsapiroutesfinetuninggetfinetuningjob.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fccc29306bf7
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const JobsApiRoutesFineTuningGetFineTuningJobRequest$outboundSchema = z.object({
    jobId: z.string(),
}).transform((v) => {
    return remap$(v, {
        jobId: "job_id",
    });
});
export function jobsApiRoutesFineTuningGetFineTuningJobRequestToJSON(jobsApiRoutesFineTuningGetFineTuningJobRequest) {
    return JSON.stringify(JobsApiRoutesFineTuningGetFineTuningJobRequest$outboundSchema.parse(jobsApiRoutesFineTuningGetFineTuningJobRequest));
}
/** @internal */
export const JobsApiRoutesFineTuningGetFineTuningJobResponse$inboundSchema = discriminatedUnion("job_type", {
    classifier: components.ClassifierFineTuningJobDetails$inboundSchema,
    completion: components.CompletionFineTuningJobDetails$inboundSchema,
}, { outputPropertyName: "jobType" });
export function jobsApiRoutesFineTuningGetFineTuningJobResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JobsApiRoutesFineTuningGetFineTuningJobResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JobsApiRoutesFineTuningGetFineTuningJobResponse' from JSON`);
}
