// @mistralai/mistralai/esm/models/operations/jobsapiroutesfinetuninggetfinetuningjobs.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 06439aebd08c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/**
 * The current job state to filter on. When set, the other results are not displayed.
 */
export const JobsApiRoutesFineTuningGetFineTuningJobsStatus = {
    Queued: "QUEUED",
    Started: "STARTED",
    Validating: "VALIDATING",
    Validated: "VALIDATED",
    Running: "RUNNING",
    FailedValidation: "FAILED_VALIDATION",
    Failed: "FAILED",
    Success: "SUCCESS",
    Cancelled: "CANCELLED",
    CancellationRequested: "CANCELLATION_REQUESTED",
};
/** @internal */
export const JobsApiRoutesFineTuningGetFineTuningJobsStatus$outboundSchema = z.enum(JobsApiRoutesFineTuningGetFineTuningJobsStatus);
/** @internal */
export const JobsApiRoutesFineTuningGetFineTuningJobsRequest$outboundSchema = z.object({
    page: z.int().default(0),
    pageSize: z.int().default(100),
    model: z.nullable(z.string()).optional(),
    createdAfter: z.nullable(z.date().transform(v => v.toISOString()))
        .optional(),
    createdBefore: z.nullable(z.date().transform(v => v.toISOString()))
        .optional(),
    createdByMe: z.boolean().default(false),
    status: z.nullable(JobsApiRoutesFineTuningGetFineTuningJobsStatus$outboundSchema).optional(),
    wandbProject: z.nullable(z.string()).optional(),
    wandbName: z.nullable(z.string()).optional(),
    suffix: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        createdAfter: "created_after",
        createdBefore: "created_before",
        createdByMe: "created_by_me",
        wandbProject: "wandb_project",
        wandbName: "wandb_name",
    });
});
export function jobsApiRoutesFineTuningGetFineTuningJobsRequestToJSON(jobsApiRoutesFineTuningGetFineTuningJobsRequest) {
    return JSON.stringify(JobsApiRoutesFineTuningGetFineTuningJobsRequest$outboundSchema.parse(jobsApiRoutesFineTuningGetFineTuningJobsRequest));
}
