// @mistralai/mistralai/esm/models/operations/jobsapiroutesbatchgetbatchjobs.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5f4c1681ccd8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";
export const OrderBy = {
    Created: "created",
    MinusCreated: "-created",
};
/** @internal */
export const OrderBy$outboundSchema = z.enum(OrderBy);
/** @internal */
export const JobsApiRoutesBatchGetBatchJobsRequest$outboundSchema = z.object({
    page: z.int().default(0),
    pageSize: z.int().default(100),
    model: z.nullable(z.string()).optional(),
    agentId: z.nullable(z.string()).optional(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    createdAfter: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    createdByMe: z.boolean().default(false),
    status: z.nullable(z.array(components.BatchJobStatus$outboundSchema))
        .optional(),
    orderBy: OrderBy$outboundSchema.default("-created"),
}).transform((v) => {
    return remap$(v, {
        pageSize: "page_size",
        agentId: "agent_id",
        createdAfter: "created_after",
        createdByMe: "created_by_me",
        orderBy: "order_by",
    });
});
export function jobsApiRoutesBatchGetBatchJobsRequestToJSON(jobsApiRoutesBatchGetBatchJobsRequest) {
    return JSON.stringify(JobsApiRoutesBatchGetBatchJobsRequest$outboundSchema.parse(jobsApiRoutesBatchGetBatchJobsRequest));
}
