// @mistralai/mistralai/esm/models/operations/jobsapiroutesfinetuningcancelfinetuningjob.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 252d6c0d8fdf
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const JobsApiRoutesFineTuningCancelFineTuningJobRequest$outboundSchema = z.object({
    jobId: z.string(),
}).transform((v) => {
    return remap$(v, {
        jobId: "job_id",
    });
});
export function jobsApiRoutesFineTuningCancelFineTuningJobRequestToJSON(jobsApiRoutesFineTuningCancelFineTuningJobRequest) {
    return JSON.stringify(JobsApiRoutesFineTuningCancelFineTuningJobRequest$outboundSchema.parse(jobsApiRoutesFineTuningCancelFineTuningJobRequest));
}
/** @internal */
export const JobsApiRoutesFineTuningCancelFineTuningJobResponse$inboundSchema = discriminatedUnion("job_type", {
    classifier: components.ClassifierFineTuningJobDetails$inboundSchema,
    completion: components.CompletionFineTuningJobDetails$inboundSchema,
}, { outputPropertyName: "jobType" });
export function jobsApiRoutesFineTuningCancelFineTuningJobResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JobsApiRoutesFineTuningCancelFineTuningJobResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JobsApiRoutesFineTuningCancelFineTuningJobResponse' from JSON`);
}
