// @mistralai/mistralai/esm/models/operations/jobsapiroutesfinetuningstartfinetuningjob.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0206a2b9695d
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const JobsApiRoutesFineTuningStartFineTuningJobRequest$outboundSchema = z.object({
    jobId: z.string(),
}).transform((v) => {
    return remap$(v, {
        jobId: "job_id",
    });
});
export function jobsApiRoutesFineTuningStartFineTuningJobRequestToJSON(jobsApiRoutesFineTuningStartFineTuningJobRequest) {
    return JSON.stringify(JobsApiRoutesFineTuningStartFineTuningJobRequest$outboundSchema.parse(jobsApiRoutesFineTuningStartFineTuningJobRequest));
}
/** @internal */
export const JobsApiRoutesFineTuningStartFineTuningJobResponse$inboundSchema = discriminatedUnion("job_type", {
    classifier: components.ClassifierFineTuningJobDetails$inboundSchema,
    completion: components.CompletionFineTuningJobDetails$inboundSchema,
}, { outputPropertyName: "jobType" });
export function jobsApiRoutesFineTuningStartFineTuningJobResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JobsApiRoutesFineTuningStartFineTuningJobResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JobsApiRoutesFineTuningStartFineTuningJobResponse' from JSON`);
}
