// @mistralai/mistralai/esm/models/operations/librariesdocumentslistv1.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f85fc1ec404d
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/** @internal */
export const LibrariesDocumentsListV1Request$outboundSchema = z.object({
    libraryId: z.string(),
    search: z.nullable(z.string()).optional(),
    pageSize: z.int().default(100),
    page: z.int().default(0),
    filtersAttributes: z.nullable(z.string()).optional(),
    sortBy: z.string().default("created_at"),
    sortOrder: z.string().default("desc"),
}).transform((v) => {
    return remap$(v, {
        libraryId: "library_id",
        pageSize: "page_size",
        filtersAttributes: "filters_attributes",
        sortBy: "sort_by",
        sortOrder: "sort_order",
    });
});
export function librariesDocumentsListV1RequestToJSON(librariesDocumentsListV1Request) {
    return JSON.stringify(LibrariesDocumentsListV1Request$outboundSchema.parse(librariesDocumentsListV1Request));
}
