// @mistralai/mistralai/esm/models/operations/librariesdocumentsuploadv1.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 55edfa30408e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { blobLikeSchema } from "../../types/blobs.js";
import * as components from "../components/index.js";
/** @internal */
export const DocumentUpload$outboundSchema = z.object({
    file: components.FileT$outboundSchema.or(blobLikeSchema),
});
export function documentUploadToJSON(documentUpload) {
    return JSON.stringify(DocumentUpload$outboundSchema.parse(documentUpload));
}
/** @internal */
export const LibrariesDocumentsUploadV1Request$outboundSchema = z.object({
    libraryId: z.string(),
    requestBody: z.lazy(() => DocumentUpload$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        libraryId: "library_id",
        requestBody: "RequestBody",
    });
});
export function librariesDocumentsUploadV1RequestToJSON(librariesDocumentsUploadV1Request) {
    return JSON.stringify(LibrariesDocumentsUploadV1Request$outboundSchema.parse(librariesDocumentsUploadV1Request));
}
