// @mistralai/mistralai/esm/models/operations/retrievemodelv1modelsmodelidget.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 723c6abd4e70
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import * as components from "../components/index.js";
/** @internal */
export const RetrieveModelV1ModelsModelIdGetRequest$outboundSchema = z.object({
    modelId: z.string(),
}).transform((v) => {
    return remap$(v, {
        modelId: "model_id",
    });
});
export function retrieveModelV1ModelsModelIdGetRequestToJSON(retrieveModelV1ModelsModelIdGetRequest) {
    return JSON.stringify(RetrieveModelV1ModelsModelIdGetRequest$outboundSchema.parse(retrieveModelV1ModelsModelIdGetRequest));
}
/** @internal */
export const ResponseRetrieveModelV1ModelsModelIdGet$inboundSchema = discriminatedUnion("type", {
    base: components.BaseModelCard$inboundSchema,
    ["fine-tuned"]: components.FTModelCard$inboundSchema,
});
export function responseRetrieveModelV1ModelsModelIdGetFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ResponseRetrieveModelV1ModelsModelIdGet$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ResponseRetrieveModelV1ModelsModelIdGet' from JSON`);
}
