// @mistralai/mistralai/esm/sdk/campaigns.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 026c94904537
 */
import { betaObservabilityCampaignsCreate } from "../funcs/betaObservabilityCampaignsCreate.js";
import { betaObservabilityCampaignsDelete } from "../funcs/betaObservabilityCampaignsDelete.js";
import { betaObservabilityCampaignsFetch } from "../funcs/betaObservabilityCampaignsFetch.js";
import { betaObservabilityCampaignsFetchStatus } from "../funcs/betaObservabilityCampaignsFetchStatus.js";
import { betaObservabilityCampaignsList } from "../funcs/betaObservabilityCampaignsList.js";
import { betaObservabilityCampaignsListEvents } from "../funcs/betaObservabilityCampaignsListEvents.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Campaigns extends ClientSDK {
    /**
     * Create and start a new campaign
     */
    async create(request, options) {
        return unwrapAsync(betaObservabilityCampaignsCreate(this, request, options));
    }
    /**
     * Get all campaigns
     */
    async list(request, options) {
        return unwrapAsync(betaObservabilityCampaignsList(this, request, options));
    }
    /**
     * Get campaign by id
     */
    async fetch(request, options) {
        return unwrapAsync(betaObservabilityCampaignsFetch(this, request, options));
    }
    /**
     * Delete a campaign
     */
    async delete(request, options) {
        return unwrapAsync(betaObservabilityCampaignsDelete(this, request, options));
    }
    /**
     * Get campaign status by campaign id
     */
    async fetchStatus(request, options) {
        return unwrapAsync(betaObservabilityCampaignsFetchStatus(this, request, options));
    }
    /**
     * Get event ids that were selected by the given campaign
     */
    async listEvents(request, options) {
        return unwrapAsync(betaObservabilityCampaignsListEvents(this, request, options));
    }
}
