// @mistralai/mistralai/esm/sdk/chatcompletionevents.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a73bb90ada41
 */
import { betaObservabilityChatCompletionEventsFetch } from "../funcs/betaObservabilityChatCompletionEventsFetch.js";
import { betaObservabilityChatCompletionEventsFetchSimilarEvents } from "../funcs/betaObservabilityChatCompletionEventsFetchSimilarEvents.js";
import { betaObservabilityChatCompletionEventsJudge } from "../funcs/betaObservabilityChatCompletionEventsJudge.js";
import { betaObservabilityChatCompletionEventsSearch } from "../funcs/betaObservabilityChatCompletionEventsSearch.js";
import { betaObservabilityChatCompletionEventsSearchIds } from "../funcs/betaObservabilityChatCompletionEventsSearchIds.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
import { Fields } from "./fields.js";
export class ChatCompletionEvents extends ClientSDK {
    _fields;
    get fields() {
        return (this._fields ?? (this._fields = new Fields(this._options)));
    }
    /**
     * Get Chat Completion Events
     */
    async search(request, options) {
        return unwrapAsync(betaObservabilityChatCompletionEventsSearch(this, request, options));
    }
    /**
     * Alternative to /search that returns only the IDs and that can return many IDs at once
     */
    async searchIds(request, options) {
        return unwrapAsync(betaObservabilityChatCompletionEventsSearchIds(this, request, options));
    }
    /**
     * Get Chat Completion Event
     */
    async fetch(request, options) {
        return unwrapAsync(betaObservabilityChatCompletionEventsFetch(this, request, options));
    }
    /**
     * Get Similar Chat Completion Events
     */
    async fetchSimilarEvents(request, options) {
        return unwrapAsync(betaObservabilityChatCompletionEventsFetchSimilarEvents(this, request, options));
    }
    /**
     * Run Judge on an event based on the given options
     */
    async judge(request, options) {
        return unwrapAsync(betaObservabilityChatCompletionEventsJudge(this, request, options));
    }
}
