// @mistralai/mistralai/esm/sdk/connectors.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2dda252d574b
 */
import { betaConnectorsActivateForOrganization } from "../funcs/betaConnectorsActivateForOrganization.js";
import { betaConnectorsActivateForUser } from "../funcs/betaConnectorsActivateForUser.js";
import { betaConnectorsActivateForWorkspace } from "../funcs/betaConnectorsActivateForWorkspace.js";
import { betaConnectorsCallTool } from "../funcs/betaConnectorsCallTool.js";
import { betaConnectorsCreate } from "../funcs/betaConnectorsCreate.js";
import { betaConnectorsCreateOrUpdateOrganizationCredentials } from "../funcs/betaConnectorsCreateOrUpdateOrganizationCredentials.js";
import { betaConnectorsCreateOrUpdateUserCredentials } from "../funcs/betaConnectorsCreateOrUpdateUserCredentials.js";
import { betaConnectorsCreateOrUpdateWorkspaceCredentials } from "../funcs/betaConnectorsCreateOrUpdateWorkspaceCredentials.js";
import { betaConnectorsDeactivateForOrganization } from "../funcs/betaConnectorsDeactivateForOrganization.js";
import { betaConnectorsDeactivateForUser } from "../funcs/betaConnectorsDeactivateForUser.js";
import { betaConnectorsDeactivateForWorkspace } from "../funcs/betaConnectorsDeactivateForWorkspace.js";
import { betaConnectorsDelete } from "../funcs/betaConnectorsDelete.js";
import { betaConnectorsDeleteOrganizationCredentials } from "../funcs/betaConnectorsDeleteOrganizationCredentials.js";
import { betaConnectorsDeleteUserCredentials } from "../funcs/betaConnectorsDeleteUserCredentials.js";
import { betaConnectorsDeleteWorkspaceCredentials } from "../funcs/betaConnectorsDeleteWorkspaceCredentials.js";
import { betaConnectorsGet } from "../funcs/betaConnectorsGet.js";
import { betaConnectorsGetAuthenticationMethods } from "../funcs/betaConnectorsGetAuthenticationMethods.js";
import { betaConnectorsGetAuthUrl } from "../funcs/betaConnectorsGetAuthUrl.js";
import { betaConnectorsList } from "../funcs/betaConnectorsList.js";
import { betaConnectorsListOrganizationCredentials } from "../funcs/betaConnectorsListOrganizationCredentials.js";
import { betaConnectorsListTools } from "../funcs/betaConnectorsListTools.js";
import { betaConnectorsListUserCredentials } from "../funcs/betaConnectorsListUserCredentials.js";
import { betaConnectorsListWorkspaceCredentials } from "../funcs/betaConnectorsListWorkspaceCredentials.js";
import { betaConnectorsUpdate } from "../funcs/betaConnectorsUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Connectors extends ClientSDK {
    /**
     * Create a new connector.
     *
     * @remarks
     * Create a new MCP connector. You can customize its visibility, url and auth type.
     */
    async create(request, options) {
        return unwrapAsync(betaConnectorsCreate(this, request, options));
    }
    /**
     * List all connectors.
     *
     * @remarks
     * List all your custom connectors with keyset pagination and filters.
     */
    async list(request, options) {
        return unwrapAsync(betaConnectorsList(this, request, options));
    }
    /**
     * Get the auth URL for a connector.
     *
     * @remarks
     * Get the OAuth2 authorization URL for a connector to initiate user authentication.
     */
    async getAuthUrl(request, options) {
        return unwrapAsync(betaConnectorsGetAuthUrl(this, request, options));
    }
    /**
     * Activate a connector for an organization.
     *
     * @remarks
     * Enable a connector at the organization level so all members can use it.
     */
    async activateForOrganization(request, options) {
        return unwrapAsync(betaConnectorsActivateForOrganization(this, request, options));
    }
    /**
     * Deactivate a connector for an organization.
     *
     * @remarks
     * Disable a connector at the organization level.
     */
    async deactivateForOrganization(request, options) {
        return unwrapAsync(betaConnectorsDeactivateForOrganization(this, request, options));
    }
    /**
     * Activate a connector for a workspace.
     *
     * @remarks
     * Enable a connector at the workspace level so all members of the workspace can use it.
     */
    async activateForWorkspace(request, options) {
        return unwrapAsync(betaConnectorsActivateForWorkspace(this, request, options));
    }
    /**
     * Deactivate a connector for a workspace.
     *
     * @remarks
     * Disable a connector at the workspace level.
     */
    async deactivateForWorkspace(request, options) {
        return unwrapAsync(betaConnectorsDeactivateForWorkspace(this, request, options));
    }
    /**
     * Activate a connector for the current user.
     *
     * @remarks
     * Enable a connector for the calling user only.
     */
    async activateForUser(request, options) {
        return unwrapAsync(betaConnectorsActivateForUser(this, request, options));
    }
    /**
     * Deactivate a connector for the current user.
     *
     * @remarks
     * Disable a connector for the calling user only.
     */
    async deactivateForUser(request, options) {
        return unwrapAsync(betaConnectorsDeactivateForUser(this, request, options));
    }
    /**
     * Call Connector Tool
     *
     * @remarks
     * Call a tool on an MCP connector.
     */
    async callTool(request, options) {
        return unwrapAsync(betaConnectorsCallTool(this, request, options));
    }
    /**
     * List tools for a connector.
     *
     * @remarks
     * List all tools available for an MCP connector.
     */
    async listTools(request, options) {
        return unwrapAsync(betaConnectorsListTools(this, request, options));
    }
    /**
     * Get authentication methods for a connector.
     *
     * @remarks
     * Get the authentication schema for a connector. Returns the list of supported authentication methods and their required headers.
     */
    async getAuthenticationMethods(request, options) {
        return unwrapAsync(betaConnectorsGetAuthenticationMethods(this, request, options));
    }
    /**
     * List organization credentials for a connector.
     *
     * @remarks
     * List all credentials configured at the organization level for a given connector.
     */
    async listOrganizationCredentials(request, options) {
        return unwrapAsync(betaConnectorsListOrganizationCredentials(this, request, options));
    }
    /**
     * Create or update organization credentials for a connector.
     *
     * @remarks
     * Create or update credentials at the organization level for a given connector.
     */
    async createOrUpdateOrganizationCredentials(request, options) {
        return unwrapAsync(betaConnectorsCreateOrUpdateOrganizationCredentials(this, request, options));
    }
    /**
     * List workspace credentials for a connector.
     *
     * @remarks
     * List all credentials configured at the workspace level for a given connector.
     */
    async listWorkspaceCredentials(request, options) {
        return unwrapAsync(betaConnectorsListWorkspaceCredentials(this, request, options));
    }
    /**
     * Create or update workspace credentials for a connector.
     *
     * @remarks
     * Create or update credentials at the workspace level for a given connector.
     */
    async createOrUpdateWorkspaceCredentials(request, options) {
        return unwrapAsync(betaConnectorsCreateOrUpdateWorkspaceCredentials(this, request, options));
    }
    /**
     * List user credentials for a connector.
     *
     * @remarks
     * List all credentials configured at the user level for a given connector.
     */
    async listUserCredentials(request, options) {
        return unwrapAsync(betaConnectorsListUserCredentials(this, request, options));
    }
    /**
     * Create or update user credentials for a connector.
     *
     * @remarks
     * Create or update credentials at the user level for a given connector.
     */
    async createOrUpdateUserCredentials(request, options) {
        return unwrapAsync(betaConnectorsCreateOrUpdateUserCredentials(this, request, options));
    }
    /**
     * Delete organization credentials for a connector.
     *
     * @remarks
     * Delete credentials at the organization level for a given connector.
     */
    async deleteOrganizationCredentials(request, options) {
        return unwrapAsync(betaConnectorsDeleteOrganizationCredentials(this, request, options));
    }
    /**
     * Delete workspace credentials for a connector.
     *
     * @remarks
     * Delete credentials at the workspace level for a given connector.
     */
    async deleteWorkspaceCredentials(request, options) {
        return unwrapAsync(betaConnectorsDeleteWorkspaceCredentials(this, request, options));
    }
    /**
     * Delete user credentials for a connector.
     *
     * @remarks
     * Delete credentials at the user level for a given connector.
     */
    async deleteUserCredentials(request, options) {
        return unwrapAsync(betaConnectorsDeleteUserCredentials(this, request, options));
    }
    /**
     * Get a connector.
     *
     * @remarks
     * Get a connector by its ID or name.
     */
    async get(request, options) {
        return unwrapAsync(betaConnectorsGet(this, request, options));
    }
    /**
     * Update a connector.
     *
     * @remarks
     * Update a connector by its ID.
     */
    async update(request, options) {
        return unwrapAsync(betaConnectorsUpdate(this, request, options));
    }
    /**
     * Delete a connector.
     *
     * @remarks
     * Delete a connector by its ID.
     */
    async delete(request, options) {
        return unwrapAsync(betaConnectorsDelete(this, request, options));
    }
}
