// @mistralai/mistralai/esm/sdk/datasets.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2f1e89a21914
 */
import { betaObservabilityDatasetsCreate } from "../funcs/betaObservabilityDatasetsCreate.js";
import { betaObservabilityDatasetsCreateRecord } from "../funcs/betaObservabilityDatasetsCreateRecord.js";
import { betaObservabilityDatasetsDelete } from "../funcs/betaObservabilityDatasetsDelete.js";
import { betaObservabilityDatasetsExportToJsonl } from "../funcs/betaObservabilityDatasetsExportToJsonl.js";
import { betaObservabilityDatasetsFetch } from "../funcs/betaObservabilityDatasetsFetch.js";
import { betaObservabilityDatasetsFetchTask } from "../funcs/betaObservabilityDatasetsFetchTask.js";
import { betaObservabilityDatasetsImportFromCampaign } from "../funcs/betaObservabilityDatasetsImportFromCampaign.js";
import { betaObservabilityDatasetsImportFromDatasetRecords } from "../funcs/betaObservabilityDatasetsImportFromDatasetRecords.js";
import { betaObservabilityDatasetsImportFromExplorer } from "../funcs/betaObservabilityDatasetsImportFromExplorer.js";
import { betaObservabilityDatasetsImportFromFile } from "../funcs/betaObservabilityDatasetsImportFromFile.js";
import { betaObservabilityDatasetsImportFromPlayground } from "../funcs/betaObservabilityDatasetsImportFromPlayground.js";
import { betaObservabilityDatasetsList } from "../funcs/betaObservabilityDatasetsList.js";
import { betaObservabilityDatasetsListRecords } from "../funcs/betaObservabilityDatasetsListRecords.js";
import { betaObservabilityDatasetsListTasks } from "../funcs/betaObservabilityDatasetsListTasks.js";
import { betaObservabilityDatasetsUpdate } from "../funcs/betaObservabilityDatasetsUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
import { Records } from "./records.js";
export class Datasets extends ClientSDK {
    _records;
    get records() {
        return (this._records ?? (this._records = new Records(this._options)));
    }
    /**
     * Create a new empty dataset
     */
    async create(request, options) {
        return unwrapAsync(betaObservabilityDatasetsCreate(this, request, options));
    }
    /**
     * List existing datasets
     */
    async list(request, options) {
        return unwrapAsync(betaObservabilityDatasetsList(this, request, options));
    }
    /**
     * Get dataset by id
     */
    async fetch(request, options) {
        return unwrapAsync(betaObservabilityDatasetsFetch(this, request, options));
    }
    /**
     * Delete a dataset
     */
    async delete(request, options) {
        return unwrapAsync(betaObservabilityDatasetsDelete(this, request, options));
    }
    /**
     * Patch dataset
     */
    async update(request, options) {
        return unwrapAsync(betaObservabilityDatasetsUpdate(this, request, options));
    }
    /**
     * List existing records in the dataset
     */
    async listRecords(request, options) {
        return unwrapAsync(betaObservabilityDatasetsListRecords(this, request, options));
    }
    /**
     * Add a conversation to the dataset
     */
    async createRecord(request, options) {
        return unwrapAsync(betaObservabilityDatasetsCreateRecord(this, request, options));
    }
    /**
     * Populate the dataset with a campaign
     */
    async importFromCampaign(request, options) {
        return unwrapAsync(betaObservabilityDatasetsImportFromCampaign(this, request, options));
    }
    /**
     * Populate the dataset with samples from the explorer
     */
    async importFromExplorer(request, options) {
        return unwrapAsync(betaObservabilityDatasetsImportFromExplorer(this, request, options));
    }
    /**
     * Populate the dataset with samples from an uploaded file
     */
    async importFromFile(request, options) {
        return unwrapAsync(betaObservabilityDatasetsImportFromFile(this, request, options));
    }
    /**
     * Populate the dataset with samples from the playground
     */
    async importFromPlayground(request, options) {
        return unwrapAsync(betaObservabilityDatasetsImportFromPlayground(this, request, options));
    }
    /**
     * Populate the dataset with samples from another dataset
     */
    async importFromDatasetRecords(request, options) {
        return unwrapAsync(betaObservabilityDatasetsImportFromDatasetRecords(this, request, options));
    }
    /**
     * Export to the Files API and retrieve presigned URL to download the resulting JSONL file
     */
    async exportToJsonl(request, options) {
        return unwrapAsync(betaObservabilityDatasetsExportToJsonl(this, request, options));
    }
    /**
     * Get status of a dataset import task
     */
    async fetchTask(request, options) {
        return unwrapAsync(betaObservabilityDatasetsFetchTask(this, request, options));
    }
    /**
     * List import tasks for the given dataset
     */
    async listTasks(request, options) {
        return unwrapAsync(betaObservabilityDatasetsListTasks(this, request, options));
    }
}
