// @mistralai/mistralai/esm/sdk/executions.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 72bfd98dad7d
 */
import { workflowsExecutionsBatchCancelWorkflowExecutions } from "../funcs/workflowsExecutionsBatchCancelWorkflowExecutions.js";
import { workflowsExecutionsBatchTerminateWorkflowExecutions } from "../funcs/workflowsExecutionsBatchTerminateWorkflowExecutions.js";
import { workflowsExecutionsCancelWorkflowExecution } from "../funcs/workflowsExecutionsCancelWorkflowExecution.js";
import { workflowsExecutionsGetWorkflowExecution } from "../funcs/workflowsExecutionsGetWorkflowExecution.js";
import { workflowsExecutionsGetWorkflowExecutionHistory } from "../funcs/workflowsExecutionsGetWorkflowExecutionHistory.js";
import { workflowsExecutionsGetWorkflowExecutionTraceEvents } from "../funcs/workflowsExecutionsGetWorkflowExecutionTraceEvents.js";
import { workflowsExecutionsGetWorkflowExecutionTraceOtel } from "../funcs/workflowsExecutionsGetWorkflowExecutionTraceOtel.js";
import { workflowsExecutionsGetWorkflowExecutionTraceSummary } from "../funcs/workflowsExecutionsGetWorkflowExecutionTraceSummary.js";
import { workflowsExecutionsQueryWorkflowExecution } from "../funcs/workflowsExecutionsQueryWorkflowExecution.js";
import { workflowsExecutionsResetWorkflow } from "../funcs/workflowsExecutionsResetWorkflow.js";
import { workflowsExecutionsSignalWorkflowExecution } from "../funcs/workflowsExecutionsSignalWorkflowExecution.js";
import { workflowsExecutionsStream } from "../funcs/workflowsExecutionsStream.js";
import { workflowsExecutionsTerminateWorkflowExecution } from "../funcs/workflowsExecutionsTerminateWorkflowExecution.js";
import { workflowsExecutionsUpdateWorkflowExecution } from "../funcs/workflowsExecutionsUpdateWorkflowExecution.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Executions extends ClientSDK {
    /**
     * Get Workflow Execution
     */
    async getWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsGetWorkflowExecution(this, request, options));
    }
    /**
     * Get Workflow Execution History
     */
    async getWorkflowExecutionHistory(request, options) {
        return unwrapAsync(workflowsExecutionsGetWorkflowExecutionHistory(this, request, options));
    }
    /**
     * Signal Workflow Execution
     */
    async signalWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsSignalWorkflowExecution(this, request, options));
    }
    /**
     * Query Workflow Execution
     */
    async queryWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsQueryWorkflowExecution(this, request, options));
    }
    /**
     * Terminate Workflow Execution
     */
    async terminateWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsTerminateWorkflowExecution(this, request, options));
    }
    /**
     * Batch Terminate Workflow Executions
     */
    async batchTerminateWorkflowExecutions(request, options) {
        return unwrapAsync(workflowsExecutionsBatchTerminateWorkflowExecutions(this, request, options));
    }
    /**
     * Cancel Workflow Execution
     */
    async cancelWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsCancelWorkflowExecution(this, request, options));
    }
    /**
     * Batch Cancel Workflow Executions
     */
    async batchCancelWorkflowExecutions(request, options) {
        return unwrapAsync(workflowsExecutionsBatchCancelWorkflowExecutions(this, request, options));
    }
    /**
     * Reset Workflow
     */
    async resetWorkflow(request, options) {
        return unwrapAsync(workflowsExecutionsResetWorkflow(this, request, options));
    }
    /**
     * Update Workflow Execution
     */
    async updateWorkflowExecution(request, options) {
        return unwrapAsync(workflowsExecutionsUpdateWorkflowExecution(this, request, options));
    }
    /**
     * Get Workflow Execution Trace Otel
     */
    async getWorkflowExecutionTraceOtel(request, options) {
        return unwrapAsync(workflowsExecutionsGetWorkflowExecutionTraceOtel(this, request, options));
    }
    /**
     * Get Workflow Execution Trace Summary
     */
    async getWorkflowExecutionTraceSummary(request, options) {
        return unwrapAsync(workflowsExecutionsGetWorkflowExecutionTraceSummary(this, request, options));
    }
    /**
     * Get Workflow Execution Trace Events
     */
    async getWorkflowExecutionTraceEvents(request, options) {
        return unwrapAsync(workflowsExecutionsGetWorkflowExecutionTraceEvents(this, request, options));
    }
    /**
     * Stream
     */
    async stream(request, options) {
        return unwrapAsync(workflowsExecutionsStream(this, request, options));
    }
}
