// @mistralai/mistralai/esm/sdk/files.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: bb9172f41ecb
 */
import { filesDelete } from "../funcs/filesDelete.js";
import { filesDownload } from "../funcs/filesDownload.js";
import { filesGetSignedUrl } from "../funcs/filesGetSignedUrl.js";
import { filesList } from "../funcs/filesList.js";
import { filesRetrieve } from "../funcs/filesRetrieve.js";
import { filesUpload } from "../funcs/filesUpload.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Files extends ClientSDK {
    /**
     * Upload File
     *
     * @remarks
     * Upload a file that can be used across various endpoints.
     *
     * The size of individual files can be a maximum of 512 MB. The Fine-tuning API only supports .jsonl files.
     *
     * Please contact us if you need to increase these storage limits.
     */
    async upload(request, options) {
        return unwrapAsync(filesUpload(this, request, options));
    }
    /**
     * List Files
     *
     * @remarks
     * Returns a list of files that belong to the user's organization.
     */
    async list(request, options) {
        return unwrapAsync(filesList(this, request, options));
    }
    /**
     * Retrieve File
     *
     * @remarks
     * Returns information about a specific file.
     */
    async retrieve(request, options) {
        return unwrapAsync(filesRetrieve(this, request, options));
    }
    /**
     * Delete File
     *
     * @remarks
     * Delete a file.
     */
    async delete(request, options) {
        return unwrapAsync(filesDelete(this, request, options));
    }
    /**
     * Download File
     *
     * @remarks
     * Download a file
     */
    async download(request, options) {
        return unwrapAsync(filesDownload(this, request, options));
    }
    /**
     * Get Signed Url
     */
    async getSignedUrl(request, options) {
        return unwrapAsync(filesGetSignedUrl(this, request, options));
    }
}
