// @mistralai/mistralai/esm/sdk/judges.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4cc54a0b1347
 */
import { betaObservabilityJudgesCreate } from "../funcs/betaObservabilityJudgesCreate.js";
import { betaObservabilityJudgesDelete } from "../funcs/betaObservabilityJudgesDelete.js";
import { betaObservabilityJudgesFetch } from "../funcs/betaObservabilityJudgesFetch.js";
import { betaObservabilityJudgesJudgeConversation } from "../funcs/betaObservabilityJudgesJudgeConversation.js";
import { betaObservabilityJudgesList } from "../funcs/betaObservabilityJudgesList.js";
import { betaObservabilityJudgesUpdate } from "../funcs/betaObservabilityJudgesUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Judges extends ClientSDK {
    /**
     * Create a new judge
     */
    async create(request, options) {
        return unwrapAsync(betaObservabilityJudgesCreate(this, request, options));
    }
    /**
     * Get judges with optional filtering and search
     */
    async list(request, options) {
        return unwrapAsync(betaObservabilityJudgesList(this, request, options));
    }
    /**
     * Get judge by id
     */
    async fetch(request, options) {
        return unwrapAsync(betaObservabilityJudgesFetch(this, request, options));
    }
    /**
     * Delete a judge
     */
    async delete(request, options) {
        return unwrapAsync(betaObservabilityJudgesDelete(this, request, options));
    }
    /**
     * Update a judge
     */
    async update(request, options) {
        return unwrapAsync(betaObservabilityJudgesUpdate(this, request, options));
    }
    /**
     * Run a saved judge on a conversation
     */
    async judgeConversation(request, options) {
        return unwrapAsync(betaObservabilityJudgesJudgeConversation(this, request, options));
    }
}
