// @mistralai/mistralai/esm/sdk/libraries.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 22660efa17a7
 */
import { betaLibrariesCreate } from "../funcs/betaLibrariesCreate.js";
import { betaLibrariesDelete } from "../funcs/betaLibrariesDelete.js";
import { betaLibrariesGet } from "../funcs/betaLibrariesGet.js";
import { betaLibrariesLibrariesUpdateV1 } from "../funcs/betaLibrariesLibrariesUpdateV1.js";
import { betaLibrariesList } from "../funcs/betaLibrariesList.js";
import { betaLibrariesUpdate } from "../funcs/betaLibrariesUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
import { Accesses } from "./accesses.js";
import { Documents } from "./documents.js";
export class Libraries extends ClientSDK {
    _documents;
    get documents() {
        return (this._documents ?? (this._documents = new Documents(this._options)));
    }
    _accesses;
    get accesses() {
        return (this._accesses ?? (this._accesses = new Accesses(this._options)));
    }
    /**
     * List all libraries you have access to.
     *
     * @remarks
     * List all libraries that you have created or have been shared with you.
     */
    async list(request, options) {
        return unwrapAsync(betaLibrariesList(this, request, options));
    }
    /**
     * Create a new Library.
     *
     * @remarks
     * Create a new Library, you will be marked as the owner and only you will have the possibility to share it with others. When first created this will only be accessible by you.
     */
    async create(request, options) {
        return unwrapAsync(betaLibrariesCreate(this, request, options));
    }
    /**
     * Detailed information about a specific Library.
     *
     * @remarks
     * Given a library id, details information about that Library.
     */
    async get(request, options) {
        return unwrapAsync(betaLibrariesGet(this, request, options));
    }
    /**
     * Delete a library and all of it's document.
     *
     * @remarks
     * Given a library id, deletes it together with all documents that have been uploaded to that library. Warning: the response will change from 200 (returning the deleted library) to 204 No Content in a future version.
     */
    async delete(request, options) {
        return unwrapAsync(betaLibrariesDelete(this, request, options));
    }
    /**
     * Update a library.
     *
     * @remarks
     * Given a library id, you can update the name and description.
     */
    async update(request, options) {
        return unwrapAsync(betaLibrariesUpdate(this, request, options));
    }
    /**
     * Update a library.
     *
     * @remarks
     * Given a library id, you can update the name and description.
     *
     * @deprecated method: Use the PATCH method instead. This PUT endpoint will be removed in a future version..
     */
    async librariesUpdateV1(request, options) {
        return unwrapAsync(betaLibrariesLibrariesUpdateV1(this, request, options));
    }
}
