// @mistralai/mistralai/esm/sdk/models.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8c761ab180fd
 */
import { modelsArchive } from "../funcs/modelsArchive.js";
import { modelsDelete } from "../funcs/modelsDelete.js";
import { modelsList } from "../funcs/modelsList.js";
import { modelsRetrieve } from "../funcs/modelsRetrieve.js";
import { modelsUnarchive } from "../funcs/modelsUnarchive.js";
import { modelsUpdate } from "../funcs/modelsUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Models extends ClientSDK {
    /**
     * List Models
     *
     * @remarks
     * List all models available to the user.
     */
    async list(request, options) {
        return unwrapAsync(modelsList(this, request, options));
    }
    /**
     * Retrieve Model
     *
     * @remarks
     * Retrieve information about a model.
     */
    async retrieve(request, options) {
        return unwrapAsync(modelsRetrieve(this, request, options));
    }
    /**
     * Delete Model
     *
     * @remarks
     * Delete a fine-tuned model.
     */
    async delete(request, options) {
        return unwrapAsync(modelsDelete(this, request, options));
    }
    /**
     * Update Fine Tuned Model
     *
     * @remarks
     * Update a model name or description.
     */
    async update(request, options) {
        return unwrapAsync(modelsUpdate(this, request, options));
    }
    /**
     * Archive Fine Tuned Model
     *
     * @remarks
     * Archive a fine-tuned model.
     */
    async archive(request, options) {
        return unwrapAsync(modelsArchive(this, request, options));
    }
    /**
     * Unarchive Fine Tuned Model
     *
     * @remarks
     * Un-archive a fine-tuned model.
     */
    async unarchive(request, options) {
        return unwrapAsync(modelsUnarchive(this, request, options));
    }
}
