// @mistralai/mistralai/esm/sdk/records.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6c95f11fdf04
 */
import { betaObservabilityDatasetsRecordsBulkDelete } from "../funcs/betaObservabilityDatasetsRecordsBulkDelete.js";
import { betaObservabilityDatasetsRecordsDelete } from "../funcs/betaObservabilityDatasetsRecordsDelete.js";
import { betaObservabilityDatasetsRecordsFetch } from "../funcs/betaObservabilityDatasetsRecordsFetch.js";
import { betaObservabilityDatasetsRecordsJudge } from "../funcs/betaObservabilityDatasetsRecordsJudge.js";
import { betaObservabilityDatasetsRecordsUpdatePayload } from "../funcs/betaObservabilityDatasetsRecordsUpdatePayload.js";
import { betaObservabilityDatasetsRecordsUpdateProperties } from "../funcs/betaObservabilityDatasetsRecordsUpdateProperties.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
export class Records extends ClientSDK {
    /**
     * Get the content of a given conversation from a dataset
     */
    async fetch(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsFetch(this, request, options));
    }
    /**
     * Delete a record from a dataset
     */
    async delete(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsDelete(this, request, options));
    }
    /**
     * Delete multiple records from datasets
     */
    async bulkDelete(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsBulkDelete(this, request, options));
    }
    /**
     * Run Judge on a dataset record based on the given options
     */
    async judge(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsJudge(this, request, options));
    }
    /**
     * Update a dataset record conversation payload
     */
    async updatePayload(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsUpdatePayload(this, request, options));
    }
    /**
     * Update conversation properties
     */
    async updateProperties(request, options) {
        return unwrapAsync(betaObservabilityDatasetsRecordsUpdateProperties(this, request, options));
    }
}
