// @mistralai/mistralai/esm/sdk/schedules.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ffd650bcc566
 */
import { workflowsSchedulesGetSchedule } from "../funcs/workflowsSchedulesGetSchedule.js";
import { workflowsSchedulesGetSchedules } from "../funcs/workflowsSchedulesGetSchedules.js";
import { workflowsSchedulesPauseSchedule } from "../funcs/workflowsSchedulesPauseSchedule.js";
import { workflowsSchedulesResumeSchedule } from "../funcs/workflowsSchedulesResumeSchedule.js";
import { workflowsSchedulesScheduleWorkflow } from "../funcs/workflowsSchedulesScheduleWorkflow.js";
import { workflowsSchedulesUnscheduleWorkflow } from "../funcs/workflowsSchedulesUnscheduleWorkflow.js";
import { workflowsSchedulesUpdateSchedule } from "../funcs/workflowsSchedulesUpdateSchedule.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
import { unwrapResultIterator } from "../types/operations.js";
export class Schedules extends ClientSDK {
    /**
     * Get Schedules
     */
    async getSchedules(request, options) {
        return unwrapResultIterator(workflowsSchedulesGetSchedules(this, request, options));
    }
    /**
     * Schedule Workflow
     */
    async scheduleWorkflow(request, options) {
        return unwrapAsync(workflowsSchedulesScheduleWorkflow(this, request, options));
    }
    /**
     * Get Schedule
     */
    async getSchedule(request, options) {
        return unwrapAsync(workflowsSchedulesGetSchedule(this, request, options));
    }
    /**
     * Unschedule Workflow
     */
    async unscheduleWorkflow(request, options) {
        return unwrapAsync(workflowsSchedulesUnscheduleWorkflow(this, request, options));
    }
    /**
     * Update Schedule
     */
    async updateSchedule(request, options) {
        return unwrapAsync(workflowsSchedulesUpdateSchedule(this, request, options));
    }
    /**
     * Pause Schedule
     */
    async pauseSchedule(request, options) {
        return unwrapAsync(workflowsSchedulesPauseSchedule(this, request, options));
    }
    /**
     * Resume Schedule
     */
    async resumeSchedule(request, options) {
        return unwrapAsync(workflowsSchedulesResumeSchedule(this, request, options));
    }
}
