// @mistralai/mistralai/esm/sdk/sdk.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 784571af2f69
 */
import { ClientSDK } from "../lib/sdks.js";
import { Agents } from "./agents.js";
import { Audio } from "./audio.js";
import { Batch } from "./batch.js";
import { Beta } from "./beta.js";
import { Chat } from "./chat.js";
import { Classifiers } from "./classifiers.js";
import { Embeddings } from "./embeddings.js";
import { Events } from "./events.js";
import { Files } from "./files.js";
import { Fim } from "./fim.js";
import { FineTuning } from "./finetuning.js";
import { Models } from "./models.js";
import { Ocr } from "./ocr.js";
import { Workflows } from "./workflows.js";
export class Mistral extends ClientSDK {
    _audio;
    get audio() {
        return (this._audio ?? (this._audio = new Audio(this._options)));
    }
    _models;
    get models() {
        return (this._models ?? (this._models = new Models(this._options)));
    }
    _beta;
    get beta() {
        return (this._beta ?? (this._beta = new Beta(this._options)));
    }
    _files;
    get files() {
        return (this._files ?? (this._files = new Files(this._options)));
    }
    _fineTuning;
    get fineTuning() {
        return (this._fineTuning ?? (this._fineTuning = new FineTuning(this._options)));
    }
    _batch;
    get batch() {
        return (this._batch ?? (this._batch = new Batch(this._options)));
    }
    _chat;
    get chat() {
        return (this._chat ?? (this._chat = new Chat(this._options)));
    }
    _fim;
    get fim() {
        return (this._fim ?? (this._fim = new Fim(this._options)));
    }
    _agents;
    get agents() {
        return (this._agents ?? (this._agents = new Agents(this._options)));
    }
    _embeddings;
    get embeddings() {
        return (this._embeddings ?? (this._embeddings = new Embeddings(this._options)));
    }
    _classifiers;
    get classifiers() {
        return (this._classifiers ?? (this._classifiers = new Classifiers(this._options)));
    }
    _ocr;
    get ocr() {
        return (this._ocr ?? (this._ocr = new Ocr(this._options)));
    }
    _workflows;
    get workflows() {
        return (this._workflows ?? (this._workflows = new Workflows(this._options)));
    }
    _events;
    get events() {
        return (this._events ?? (this._events = new Events(this._options)));
    }
}
