// @mistralai/mistralai/esm/models/components/activitytaskretryingresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 283a6c69c1b0
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ActivityTaskRetryingAttributes$inboundSchema, } from "./activitytaskretryingattributes.js";
/** @internal */
export const ActivityTaskRetryingResponse$inboundSchema = z.object({
    event_id: z.string(),
    event_timestamp: z.int(),
    root_workflow_exec_id: z.string(),
    parent_workflow_exec_id: z.nullable(z.string()),
    workflow_exec_id: z.string(),
    workflow_run_id: z.string(),
    workflow_name: z.string(),
    event_type: z.literal("ACTIVITY_TASK_RETRYING").default("ACTIVITY_TASK_RETRYING"),
    attributes: ActivityTaskRetryingAttributes$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "event_id": "eventId",
        "event_timestamp": "eventTimestamp",
        "root_workflow_exec_id": "rootWorkflowExecId",
        "parent_workflow_exec_id": "parentWorkflowExecId",
        "workflow_exec_id": "workflowExecId",
        "workflow_run_id": "workflowRunId",
        "workflow_name": "workflowName",
        "event_type": "eventType",
    });
});
export function activityTaskRetryingResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ActivityTaskRetryingResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActivityTaskRetryingResponse' from JSON`);
}
