// @mistralai/mistralai/esm/models/components/agenthandoffdoneevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2a9d3290510a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const AgentHandoffDoneEvent$inboundSchema = z.object({
    type: z.literal("agent.handoff.done"),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v))
        .optional(),
    output_index: z.int().default(0),
    id: z.string(),
    next_agent_id: z.string(),
    next_agent_name: z.string(),
}).transform((v) => {
    return remap$(v, {
        "created_at": "createdAt",
        "output_index": "outputIndex",
        "next_agent_id": "nextAgentId",
        "next_agent_name": "nextAgentName",
    });
});
export function agentHandoffDoneEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AgentHandoffDoneEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AgentHandoffDoneEvent' from JSON`);
}
