// @mistralai/mistralai/esm/models/components/agenthandoffstartedevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 058b44bbb04f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const AgentHandoffStartedEvent$inboundSchema = z.object({
    type: z.literal("agent.handoff.started"),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v))
        .optional(),
    output_index: z.int().default(0),
    id: z.string(),
    previous_agent_id: z.string(),
    previous_agent_name: z.string(),
}).transform((v) => {
    return remap$(v, {
        "created_at": "createdAt",
        "output_index": "outputIndex",
        "previous_agent_id": "previousAgentId",
        "previous_agent_name": "previousAgentName",
    });
});
export function agentHandoffStartedEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AgentHandoffStartedEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AgentHandoffStartedEvent' from JSON`);
}
