// @mistralai/mistralai/esm/models/components/audiochunk.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 561de42bd77e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const AudioChunk$inboundSchema = z
    .object({
    type: z.literal("input_audio"),
    input_audio: z.string(),
}).transform((v) => {
    return remap$(v, {
        "input_audio": "inputAudio",
    });
});
/** @internal */
export const AudioChunk$outboundSchema = z.object({
    type: z.literal("input_audio"),
    inputAudio: z.string(),
}).transform((v) => {
    return remap$(v, {
        inputAudio: "input_audio",
    });
});
export function audioChunkToJSON(audioChunk) {
    return JSON.stringify(AudioChunk$outboundSchema.parse(audioChunk));
}
export function audioChunkFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AudioChunk$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AudioChunk' from JSON`);
}
