// @mistralai/mistralai/esm/models/components/audiotranscriptionrequeststream.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d436d38ecd10
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { blobLikeSchema } from "../../types/blobs.js";
import { FileT$outboundSchema } from "./file.js";
import { TimestampGranularity$outboundSchema, } from "./timestampgranularity.js";
/** @internal */
export const AudioTranscriptionRequestStream$outboundSchema = z.object({
    model: z.string(),
    file: FileT$outboundSchema.or(blobLikeSchema).optional(),
    fileUrl: z.nullable(z.string()).optional(),
    fileId: z.nullable(z.string()).optional(),
    language: z.nullable(z.string()).optional(),
    temperature: z.nullable(z.number()).optional(),
    stream: z.literal(true).default(true),
    diarize: z.boolean().default(false),
    contextBias: z.array(z.string()).optional(),
    timestampGranularities: z.array(TimestampGranularity$outboundSchema)
        .optional(),
}).transform((v) => {
    return remap$(v, {
        fileUrl: "file_url",
        fileId: "file_id",
        contextBias: "context_bias",
        timestampGranularities: "timestamp_granularities",
    });
});
export function audioTranscriptionRequestStreamToJSON(audioTranscriptionRequestStream) {
    return JSON.stringify(AudioTranscriptionRequestStream$outboundSchema.parse(audioTranscriptionRequestStream));
}
