// @mistralai/mistralai/esm/models/components/authenticationconfiguration.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4119721938c2
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ConsumerType$inboundSchema } from "./consumertype.js";
import { OutboundAuthenticationType$inboundSchema, } from "./outboundauthenticationtype.js";
/** @internal */
export const AuthenticationConfiguration$inboundSchema = z.object({
    name: z.string(),
    authentication_type: OutboundAuthenticationType$inboundSchema,
    scope: ConsumerType$inboundSchema,
    is_default: z.boolean().default(false),
}).transform((v) => {
    return remap$(v, {
        "authentication_type": "authenticationType",
        "is_default": "isDefault",
    });
});
export function authenticationConfigurationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AuthenticationConfiguration$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AuthenticationConfiguration' from JSON`);
}
