// @mistralai/mistralai/esm/models/components/classificationrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fb4d94629408
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { smartUnion } from "../../types/smartUnion.js";
/** @internal */
export const ClassificationRequestInputs$outboundSchema = smartUnion([z.string(), z.array(z.string())]);
export function classificationRequestInputsToJSON(classificationRequestInputs) {
    return JSON.stringify(ClassificationRequestInputs$outboundSchema.parse(classificationRequestInputs));
}
/** @internal */
export const ClassificationRequest$outboundSchema = z.object({
    model: z.string(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    inputs: smartUnion([z.string(), z.array(z.string())]),
}).transform((v) => {
    return remap$(v, {
        inputs: "input",
    });
});
export function classificationRequestToJSON(classificationRequest) {
    return JSON.stringify(ClassificationRequest$outboundSchema.parse(classificationRequest));
}
