// @mistralai/mistralai/esm/models/components/classifiertrainingparameters.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c7e214f868cb
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const ClassifierTrainingParameters$inboundSchema = z.object({
    training_steps: z.nullable(z.int()).optional(),
    learning_rate: z.number().default(0.0001),
    weight_decay: z.nullable(z.number()).optional(),
    warmup_fraction: z.nullable(z.number()).optional(),
    epochs: z.nullable(z.number()).optional(),
    seq_len: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        "training_steps": "trainingSteps",
        "learning_rate": "learningRate",
        "weight_decay": "weightDecay",
        "warmup_fraction": "warmupFraction",
        "seq_len": "seqLen",
    });
});
/** @internal */
export const ClassifierTrainingParameters$outboundSchema = z.object({
    trainingSteps: z.nullable(z.int()).optional(),
    learningRate: z.number().default(0.0001),
    weightDecay: z.nullable(z.number()).optional(),
    warmupFraction: z.nullable(z.number()).optional(),
    epochs: z.nullable(z.number()).optional(),
    seqLen: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        trainingSteps: "training_steps",
        learningRate: "learning_rate",
        weightDecay: "weight_decay",
        warmupFraction: "warmup_fraction",
        seqLen: "seq_len",
    });
});
export function classifierTrainingParametersToJSON(classifierTrainingParameters) {
    return JSON.stringify(ClassifierTrainingParameters$outboundSchema.parse(classifierTrainingParameters));
}
export function classifierTrainingParametersFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ClassifierTrainingParameters$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ClassifierTrainingParameters' from JSON`);
}
