// @mistralai/mistralai/esm/models/components/classifierfinetunedmodel.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8d7af6ec165a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ClassifierTargetResult$inboundSchema, } from "./classifiertargetresult.js";
import { FineTunedModelCapabilities$inboundSchema, } from "./finetunedmodelcapabilities.js";
/** @internal */
export const ClassifierFineTunedModel$inboundSchema = z.object({
    id: z.string(),
    object: z.literal("model").default("model"),
    created: z.int(),
    owned_by: z.string(),
    workspace_id: z.string(),
    root: z.string(),
    root_version: z.string(),
    archived: z.boolean(),
    name: z.nullable(z.string()).optional(),
    description: z.nullable(z.string()).optional(),
    capabilities: FineTunedModelCapabilities$inboundSchema,
    max_context_length: z.int().default(32768),
    aliases: z.array(z.string()).optional(),
    job: z.nullable(z.string()).optional(),
    classifier_targets: z.array(ClassifierTargetResult$inboundSchema),
    model_type: z.literal("classifier"),
}).transform((v) => {
    return remap$(v, {
        "owned_by": "ownedBy",
        "workspace_id": "workspaceId",
        "root_version": "rootVersion",
        "max_context_length": "maxContextLength",
        "classifier_targets": "classifierTargets",
        "model_type": "modelType",
    });
});
export function classifierFineTunedModelFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ClassifierFineTunedModel$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ClassifierFineTunedModel' from JSON`);
}
