// @mistralai/mistralai/esm/models/components/clientcapabilities.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3fac74df5610
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { ClientTasksCapability$inboundSchema, } from "./clienttaskscapability.js";
import { ElicitationCapability$inboundSchema, } from "./elicitationcapability.js";
import { RootsCapability$inboundSchema, } from "./rootscapability.js";
import { SamplingCapability$inboundSchema, } from "./samplingcapability.js";
/** @internal */
export const ClientCapabilities$inboundSchema = z.object({
    experimental: z.nullable(z.record(z.string(), z.record(z.string(), z.any())))
        .optional(),
    sampling: z.nullable(SamplingCapability$inboundSchema).optional(),
    elicitation: z.nullable(ElicitationCapability$inboundSchema).optional(),
    roots: z.nullable(RootsCapability$inboundSchema).optional(),
    tasks: z.nullable(ClientTasksCapability$inboundSchema).optional(),
}).catchall(z.any());
export function clientCapabilitiesFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ClientCapabilities$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ClientCapabilities' from JSON`);
}
