// @mistralai/mistralai/esm/models/components/codeinterpretertool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6a03acd8a711
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ToolConfiguration$inboundSchema, ToolConfiguration$outboundSchema, } from "./toolconfiguration.js";
/** @internal */
export const CodeInterpreterTool$inboundSchema = z.object({
    tool_configuration: z.nullable(ToolConfiguration$inboundSchema).optional(),
    type: z.literal("code_interpreter"),
}).transform((v) => {
    return remap$(v, {
        "tool_configuration": "toolConfiguration",
    });
});
/** @internal */
export const CodeInterpreterTool$outboundSchema = z.object({
    toolConfiguration: z.nullable(ToolConfiguration$outboundSchema).optional(),
    type: z.literal("code_interpreter"),
}).transform((v) => {
    return remap$(v, {
        toolConfiguration: "tool_configuration",
    });
});
export function codeInterpreterToolToJSON(codeInterpreterTool) {
    return JSON.stringify(CodeInterpreterTool$outboundSchema.parse(codeInterpreterTool));
}
export function codeInterpreterToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CodeInterpreterTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CodeInterpreterTool' from JSON`);
}
