// @mistralai/mistralai/esm/models/components/conversationappendstreamrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e089da49cde8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { CompletionArgs$outboundSchema, } from "./completionargs.js";
import { ConversationInputs$outboundSchema, } from "./conversationinputs.js";
import { ToolCallConfirmation$outboundSchema, } from "./toolcallconfirmation.js";
export const ConversationAppendStreamRequestHandoffExecution = {
    Client: "client",
    Server: "server",
};
/** @internal */
export const ConversationAppendStreamRequestHandoffExecution$outboundSchema = z.enum(ConversationAppendStreamRequestHandoffExecution);
/** @internal */
export const ConversationAppendStreamRequest$outboundSchema = z.object({
    inputs: ConversationInputs$outboundSchema.optional(),
    stream: z.literal(true).default(true),
    store: z.boolean().default(true),
    handoffExecution: ConversationAppendStreamRequestHandoffExecution$outboundSchema.default("server"),
    completionArgs: CompletionArgs$outboundSchema.optional(),
    toolConfirmations: z.nullable(z.array(ToolCallConfirmation$outboundSchema))
        .optional(),
}).transform((v) => {
    return remap$(v, {
        handoffExecution: "handoff_execution",
        completionArgs: "completion_args",
        toolConfirmations: "tool_confirmations",
    });
});
export function conversationAppendStreamRequestToJSON(conversationAppendStreamRequest) {
    return JSON.stringify(ConversationAppendStreamRequest$outboundSchema.parse(conversationAppendStreamRequest));
}
