// @mistralai/mistralai/esm/models/components/conversationappendrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 09fae83dc82a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { CompletionArgs$outboundSchema, } from "./completionargs.js";
import { ConversationInputs$outboundSchema, } from "./conversationinputs.js";
import { ToolCallConfirmation$outboundSchema, } from "./toolcallconfirmation.js";
export const ConversationAppendRequestHandoffExecution = {
    Client: "client",
    Server: "server",
};
/** @internal */
export const ConversationAppendRequestHandoffExecution$outboundSchema = z.enum(ConversationAppendRequestHandoffExecution);
/** @internal */
export const ConversationAppendRequest$outboundSchema = z.object({
    inputs: ConversationInputs$outboundSchema.optional(),
    stream: z.literal(false).default(false),
    store: z.boolean().default(true),
    handoffExecution: ConversationAppendRequestHandoffExecution$outboundSchema
        .default("server"),
    completionArgs: CompletionArgs$outboundSchema.optional(),
    toolConfirmations: z.nullable(z.array(ToolCallConfirmation$outboundSchema))
        .optional(),
}).transform((v) => {
    return remap$(v, {
        handoffExecution: "handoff_execution",
        completionArgs: "completion_args",
        toolConfirmations: "tool_confirmations",
    });
});
export function conversationAppendRequestToJSON(conversationAppendRequest) {
    return JSON.stringify(ConversationAppendRequest$outboundSchema.parse(conversationAppendRequest));
}
