// @mistralai/mistralai/esm/models/components/createbatchjobrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5e267642c095
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { ApiEndpoint$outboundSchema } from "./apiendpoint.js";
import { BatchRequest$outboundSchema, } from "./batchrequest.js";
/** @internal */
export const CreateBatchJobRequest$outboundSchema = z.object({
    inputFiles: z.nullable(z.array(z.string())).optional(),
    requests: z.nullable(z.array(BatchRequest$outboundSchema)).optional(),
    endpoint: ApiEndpoint$outboundSchema,
    model: z.nullable(z.string()).optional(),
    agentId: z.nullable(z.string()).optional(),
    metadata: z.nullable(z.record(z.string(), z.string())).optional(),
    timeoutHours: z.int().default(24),
}).transform((v) => {
    return remap$(v, {
        inputFiles: "input_files",
        agentId: "agent_id",
        timeoutHours: "timeout_hours",
    });
});
export function createBatchJobRequestToJSON(createBatchJobRequest) {
    return JSON.stringify(CreateBatchJobRequest$outboundSchema.parse(createBatchJobRequest));
}
