// @mistralai/mistralai/esm/models/components/createjudgerequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 904f380447ed
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { JudgeClassificationOutput$outboundSchema, } from "./judgeclassificationoutput.js";
import { JudgeRegressionOutput$outboundSchema, } from "./judgeregressionoutput.js";
/** @internal */
export const CreateJudgeRequestOutput$outboundSchema = z.union([
    JudgeClassificationOutput$outboundSchema,
    JudgeRegressionOutput$outboundSchema,
]);
export function createJudgeRequestOutputToJSON(createJudgeRequestOutput) {
    return JSON.stringify(CreateJudgeRequestOutput$outboundSchema.parse(createJudgeRequestOutput));
}
/** @internal */
export const CreateJudgeRequest$outboundSchema = z.object({
    name: z.string(),
    description: z.string(),
    modelName: z.string(),
    output: z.union([
        JudgeClassificationOutput$outboundSchema,
        JudgeRegressionOutput$outboundSchema,
    ]),
    instructions: z.string(),
    tools: z.array(z.string()),
}).transform((v) => {
    return remap$(v, {
        modelName: "model_name",
    });
});
export function createJudgeRequestToJSON(createJudgeRequest) {
    return JSON.stringify(CreateJudgeRequest$outboundSchema.parse(createJudgeRequest));
}
