// @mistralai/mistralai/esm/models/components/createagentrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ca59b7d3ab77
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { CodeInterpreterTool$outboundSchema, } from "./codeinterpretertool.js";
import { CompletionArgs$outboundSchema, } from "./completionargs.js";
import { CustomConnector$outboundSchema, } from "./customconnector.js";
import { DocumentLibraryTool$outboundSchema, } from "./documentlibrarytool.js";
import { FunctionTool$outboundSchema, } from "./functiontool.js";
import { GuardrailConfig$outboundSchema, } from "./guardrailconfig.js";
import { ImageGenerationTool$outboundSchema, } from "./imagegenerationtool.js";
import { WebSearchPremiumTool$outboundSchema, } from "./websearchpremiumtool.js";
import { WebSearchTool$outboundSchema, } from "./websearchtool.js";
/** @internal */
export const CreateAgentRequestTool$outboundSchema = z.union([
    CodeInterpreterTool$outboundSchema,
    CustomConnector$outboundSchema,
    DocumentLibraryTool$outboundSchema,
    FunctionTool$outboundSchema,
    ImageGenerationTool$outboundSchema,
    WebSearchTool$outboundSchema,
    WebSearchPremiumTool$outboundSchema,
]);
export function createAgentRequestToolToJSON(createAgentRequestTool) {
    return JSON.stringify(CreateAgentRequestTool$outboundSchema.parse(createAgentRequestTool));
}
/** @internal */
export const CreateAgentRequest$outboundSchema = z.object({
    instructions: z.nullable(z.string()).optional(),
    tools: z.array(z.union([
        CodeInterpreterTool$outboundSchema,
        CustomConnector$outboundSchema,
        DocumentLibraryTool$outboundSchema,
        FunctionTool$outboundSchema,
        ImageGenerationTool$outboundSchema,
        WebSearchTool$outboundSchema,
        WebSearchPremiumTool$outboundSchema,
    ])).optional(),
    completionArgs: CompletionArgs$outboundSchema.optional(),
    guardrails: z.nullable(z.array(GuardrailConfig$outboundSchema)).optional(),
    model: z.string(),
    name: z.string(),
    description: z.nullable(z.string()).optional(),
    handoffs: z.nullable(z.array(z.string())).optional(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    versionMessage: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        completionArgs: "completion_args",
        versionMessage: "version_message",
    });
});
export function createAgentRequestToJSON(createAgentRequest) {
    return JSON.stringify(CreateAgentRequest$outboundSchema.parse(createAgentRequest));
}
