// @mistralai/mistralai/esm/models/components/createsearchindexinforequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 390f11ebd46a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { CreateVespaSearchIndexInfoRequest$outboundSchema, } from "./createvespasearchindexinforequest.js";
export const CreateSearchIndexInfoRequestStatus = {
    Online: "online",
    Offline: "offline",
};
/** @internal */
export const CreateSearchIndexInfoRequestStatus$outboundSchema = z.enum(CreateSearchIndexInfoRequestStatus);
/** @internal */
export const CreateSearchIndexInfoRequestIndex$outboundSchema = CreateVespaSearchIndexInfoRequest$outboundSchema;
export function createSearchIndexInfoRequestIndexToJSON(createSearchIndexInfoRequestIndex) {
    return JSON.stringify(CreateSearchIndexInfoRequestIndex$outboundSchema.parse(createSearchIndexInfoRequestIndex));
}
/** @internal */
export const CreateSearchIndexInfoRequest$outboundSchema = z.object({
    name: z.string(),
    documentCount: z.nullable(z.int()).optional(),
    status: CreateSearchIndexInfoRequestStatus$outboundSchema.default("offline"),
    index: CreateVespaSearchIndexInfoRequest$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        documentCount: "document_count",
    });
});
export function createSearchIndexInfoRequestToJSON(createSearchIndexInfoRequest) {
    return JSON.stringify(CreateSearchIndexInfoRequest$outboundSchema.parse(createSearchIndexInfoRequest));
}
