// @mistralai/mistralai/esm/models/components/customtasktimedoutresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7af631b73371
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { CustomTaskTimedOutAttributes$inboundSchema, } from "./customtasktimedoutattributes.js";
/** @internal */
export const CustomTaskTimedOutResponse$inboundSchema = z.object({
    event_id: z.string(),
    event_timestamp: z.int(),
    root_workflow_exec_id: z.string(),
    parent_workflow_exec_id: z.nullable(z.string()),
    workflow_exec_id: z.string(),
    workflow_run_id: z.string(),
    workflow_name: z.string(),
    event_type: z.literal("CUSTOM_TASK_TIMED_OUT").default("CUSTOM_TASK_TIMED_OUT"),
    attributes: CustomTaskTimedOutAttributes$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "event_id": "eventId",
        "event_timestamp": "eventTimestamp",
        "root_workflow_exec_id": "rootWorkflowExecId",
        "parent_workflow_exec_id": "parentWorkflowExecId",
        "workflow_exec_id": "workflowExecId",
        "workflow_run_id": "workflowRunId",
        "workflow_name": "workflowName",
        "event_type": "eventType",
    });
});
export function customTaskTimedOutResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CustomTaskTimedOutResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CustomTaskTimedOutResponse' from JSON`);
}
