// @mistralai/mistralai/esm/models/components/dataset.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2b053ccea202
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const Dataset$inboundSchema = z.object({
    id: z.string(),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    updated_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    deleted_at: z.nullable(z.iso.datetime({ offset: true }).transform(v => new Date(v))),
    name: z.string(),
    description: z.string(),
    owner_id: z.string(),
    workspace_id: z.string(),
}).transform((v) => {
    return remap$(v, {
        "created_at": "createdAt",
        "updated_at": "updatedAt",
        "deleted_at": "deletedAt",
        "owner_id": "ownerId",
        "workspace_id": "workspaceId",
    });
});
export function datasetFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Dataset$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Dataset' from JSON`);
}
