// @mistralai/mistralai/esm/models/components/deploymentresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 70ec10dc50b4
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { DeploymentLocation$inboundSchema, } from "./deploymentlocation.js";
/** @internal */
export const DeploymentResponse$inboundSchema = z.object({
    id: z.string(),
    name: z.string(),
    is_active: z.boolean(),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    updated_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    location: z.nullable(DeploymentLocation$inboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        "is_active": "isActive",
        "created_at": "createdAt",
        "updated_at": "updatedAt",
    });
});
export function deploymentResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DeploymentResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeploymentResponse' from JSON`);
}
