// @mistralai/mistralai/esm/models/components/embeddingrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1f5a1e8c8aa9
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { smartUnion } from "../../types/smartUnion.js";
import { EmbeddingDtype$outboundSchema, } from "./embeddingdtype.js";
import { EncodingFormat$outboundSchema, } from "./encodingformat.js";
/** @internal */
export const EmbeddingRequestInputs$outboundSchema = smartUnion([z.string(), z.array(z.string())]);
export function embeddingRequestInputsToJSON(embeddingRequestInputs) {
    return JSON.stringify(EmbeddingRequestInputs$outboundSchema.parse(embeddingRequestInputs));
}
/** @internal */
export const EmbeddingRequest$outboundSchema = z.object({
    model: z.string(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    inputs: smartUnion([z.string(), z.array(z.string())]),
    outputDimension: z.nullable(z.int()).optional(),
    outputDtype: EmbeddingDtype$outboundSchema.optional(),
    encodingFormat: EncodingFormat$outboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        inputs: "input",
        outputDimension: "output_dimension",
        outputDtype: "output_dtype",
        encodingFormat: "encoding_format",
    });
});
export function embeddingRequestToJSON(embeddingRequest) {
    return JSON.stringify(EmbeddingRequest$outboundSchema.parse(embeddingRequest));
}
