// @mistralai/mistralai/esm/models/components/jobmetadata.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 879060fa3ae2
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const JobMetadata$inboundSchema = z
    .object({
    expected_duration_seconds: z.nullable(z.int()).optional(),
    cost: z.nullable(z.number()).optional(),
    cost_currency: z.nullable(z.string()).optional(),
    train_tokens_per_step: z.nullable(z.int()).optional(),
    train_tokens: z.nullable(z.int()).optional(),
    data_tokens: z.nullable(z.int()).optional(),
    estimated_start_time: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        "expected_duration_seconds": "expectedDurationSeconds",
        "cost_currency": "costCurrency",
        "train_tokens_per_step": "trainTokensPerStep",
        "train_tokens": "trainTokens",
        "data_tokens": "dataTokens",
        "estimated_start_time": "estimatedStartTime",
    });
});
export function jobMetadataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JobMetadata$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JobMetadata' from JSON`);
}
