// @mistralai/mistralai/esm/models/components/jsonpatchappend.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b0e6eff2bea9
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import { EncryptedPatchValue$inboundSchema, } from "./encryptedpatchvalue.js";
/** @internal */
export const JSONPatchAppendValue$inboundSchema = smartUnion([EncryptedPatchValue$inboundSchema, z.string()]);
export function jsonPatchAppendValueFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JSONPatchAppendValue$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JSONPatchAppendValue' from JSON`);
}
/** @internal */
export const JSONPatchAppend$inboundSchema = z.object({
    path: z.string(),
    value: smartUnion([EncryptedPatchValue$inboundSchema, z.string()]),
    op: z.literal("append"),
});
export function jsonPatchAppendFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JSONPatchAppend$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JSONPatchAppend' from JSON`);
}
