// @mistralai/mistralai/esm/models/components/judgeclassificationoutput.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a0f6baefc6e2
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { JudgeClassificationOutputOption$inboundSchema, JudgeClassificationOutputOption$outboundSchema, } from "./judgeclassificationoutputoption.js";
/** @internal */
export const JudgeClassificationOutput$inboundSchema = z.object({
    type: z.literal("CLASSIFICATION"),
    options: z.array(JudgeClassificationOutputOption$inboundSchema),
});
/** @internal */
export const JudgeClassificationOutput$outboundSchema = z.object({
    type: z.literal("CLASSIFICATION"),
    options: z.array(JudgeClassificationOutputOption$outboundSchema),
});
export function judgeClassificationOutputToJSON(judgeClassificationOutput) {
    return JSON.stringify(JudgeClassificationOutput$outboundSchema.parse(judgeClassificationOutput));
}
export function judgeClassificationOutputFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JudgeClassificationOutput$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JudgeClassificationOutput' from JSON`);
}
