// @mistralai/mistralai/esm/models/components/judgeregressionoutput.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d517eb109b04
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const JudgeRegressionOutput$inboundSchema = z.object({
    type: z.literal("REGRESSION"),
    min: z.number().default(0),
    min_description: z.string(),
    max: z.number().default(1),
    max_description: z.string(),
}).transform((v) => {
    return remap$(v, {
        "min_description": "minDescription",
        "max_description": "maxDescription",
    });
});
/** @internal */
export const JudgeRegressionOutput$outboundSchema = z.object({
    type: z.literal("REGRESSION"),
    min: z.number().default(0),
    minDescription: z.string(),
    max: z.number().default(1),
    maxDescription: z.string(),
}).transform((v) => {
    return remap$(v, {
        minDescription: "min_description",
        maxDescription: "max_description",
    });
});
export function judgeRegressionOutputToJSON(judgeRegressionOutput) {
    return JSON.stringify(JudgeRegressionOutput$outboundSchema.parse(judgeRegressionOutput));
}
export function judgeRegressionOutputFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JudgeRegressionOutput$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JudgeRegressionOutput' from JSON`);
}
