// @mistralai/mistralai/esm/models/components/mcpprompt.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 296526242b71
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { MCPServerIcon$inboundSchema } from "./mcpservericon.js";
import { PromptArgument$inboundSchema, } from "./promptargument.js";
/** @internal */
export const MCPPrompt$inboundSchema = z.object({
    name: z.string(),
    title: z.nullable(z.string()).optional(),
    description: z.nullable(z.string()).optional(),
    arguments: z.nullable(z.array(PromptArgument$inboundSchema)).optional(),
    icons: z.nullable(z.array(MCPServerIcon$inboundSchema)).optional(),
    _meta: z.nullable(z.record(z.string(), z.any())).optional(),
}).catchall(z.any()).transform((v) => {
    return remap$(v, {
        "_meta": "meta",
    });
});
export function mcpPromptFromJSON(jsonString) {
    return safeParse(jsonString, (x) => MCPPrompt$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'MCPPrompt' from JSON`);
}
