// @mistralai/mistralai/esm/models/components/mcpserverremote.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 50e67fe6708b
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { MCPServerAuthenticationRequirement$inboundSchema, } from "./mcpserverauthenticationrequirement.js";
import { MCPServerRemoteHeader$inboundSchema, } from "./mcpserverremoteheader.js";
/**
 * Transport type
 */
export const MCPServerRemoteType = {
    StreamableHttp: "streamable-http",
    Sse: "sse",
};
/** @internal */
export const MCPServerRemoteType$inboundSchema = openEnums.inboundSchema(MCPServerRemoteType);
/** @internal */
export const MCPServerRemote$inboundSchema = z.object({
    type: MCPServerRemoteType$inboundSchema,
    url: z.string(),
    supportedProtocolVersions: z.nullable(z.array(z.string())).optional(),
    headers: z.nullable(z.array(MCPServerRemoteHeader$inboundSchema)).optional(),
    authentication: z.nullable(MCPServerAuthenticationRequirement$inboundSchema)
        .optional(),
});
export function mcpServerRemoteFromJSON(jsonString) {
    return safeParse(jsonString, (x) => MCPServerRemote$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'MCPServerRemote' from JSON`);
}
