// @mistralai/mistralai/esm/models/components/modelconversation.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e79bc3ad623a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import { CodeInterpreterTool$inboundSchema, } from "./codeinterpretertool.js";
import { CompletionArgs$inboundSchema, } from "./completionargs.js";
import { CustomConnector$inboundSchema, } from "./customconnector.js";
import { DocumentLibraryTool$inboundSchema, } from "./documentlibrarytool.js";
import { FunctionTool$inboundSchema } from "./functiontool.js";
import { GuardrailConfig$inboundSchema, } from "./guardrailconfig.js";
import { ImageGenerationTool$inboundSchema, } from "./imagegenerationtool.js";
import { WebSearchPremiumTool$inboundSchema, } from "./websearchpremiumtool.js";
import { WebSearchTool$inboundSchema } from "./websearchtool.js";
/** @internal */
export const ModelConversationTool$inboundSchema = discriminatedUnion("type", {
    code_interpreter: CodeInterpreterTool$inboundSchema,
    connector: CustomConnector$inboundSchema,
    document_library: DocumentLibraryTool$inboundSchema,
    function: FunctionTool$inboundSchema,
    image_generation: ImageGenerationTool$inboundSchema,
    web_search: WebSearchTool$inboundSchema,
    web_search_premium: WebSearchPremiumTool$inboundSchema,
});
export function modelConversationToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ModelConversationTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ModelConversationTool' from JSON`);
}
/** @internal */
export const ModelConversation$inboundSchema = z.object({
    instructions: z.nullable(z.string()).optional(),
    tools: z.array(discriminatedUnion("type", {
        code_interpreter: CodeInterpreterTool$inboundSchema,
        connector: CustomConnector$inboundSchema,
        document_library: DocumentLibraryTool$inboundSchema,
        function: FunctionTool$inboundSchema,
        image_generation: ImageGenerationTool$inboundSchema,
        web_search: WebSearchTool$inboundSchema,
        web_search_premium: WebSearchPremiumTool$inboundSchema,
    })).optional(),
    completion_args: CompletionArgs$inboundSchema.optional(),
    guardrails: z.nullable(z.array(GuardrailConfig$inboundSchema)).optional(),
    name: z.nullable(z.string()).optional(),
    description: z.nullable(z.string()).optional(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    object: z.literal("conversation").default("conversation"),
    id: z.string(),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    updated_at: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    model: z.string(),
}).transform((v) => {
    return remap$(v, {
        "completion_args": "completionArgs",
        "created_at": "createdAt",
        "updated_at": "updatedAt",
    });
});
export function modelConversationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ModelConversation$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ModelConversation' from JSON`);
}
