// @mistralai/mistralai/esm/models/components/moderationllmv1categorythresholds.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a08990f72387
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const ModerationLlmv1CategoryThresholds$inboundSchema = z.object({
    sexual: z.nullable(z.number()).optional(),
    hate_and_discrimination: z.nullable(z.number()).optional(),
    violence_and_threats: z.nullable(z.number()).optional(),
    dangerous_and_criminal_content: z.nullable(z.number()).optional(),
    selfharm: z.nullable(z.number()).optional(),
    health: z.nullable(z.number()).optional(),
    financial: z.nullable(z.number()).optional(),
    law: z.nullable(z.number()).optional(),
    pii: z.nullable(z.number()).optional(),
}).transform((v) => {
    return remap$(v, {
        "hate_and_discrimination": "hateAndDiscrimination",
        "violence_and_threats": "violenceAndThreats",
        "dangerous_and_criminal_content": "dangerousAndCriminalContent",
    });
});
/** @internal */
export const ModerationLlmv1CategoryThresholds$outboundSchema = z.object({
    sexual: z.nullable(z.number()).optional(),
    hateAndDiscrimination: z.nullable(z.number()).optional(),
    violenceAndThreats: z.nullable(z.number()).optional(),
    dangerousAndCriminalContent: z.nullable(z.number()).optional(),
    selfharm: z.nullable(z.number()).optional(),
    health: z.nullable(z.number()).optional(),
    financial: z.nullable(z.number()).optional(),
    law: z.nullable(z.number()).optional(),
    pii: z.nullable(z.number()).optional(),
}).transform((v) => {
    return remap$(v, {
        hateAndDiscrimination: "hate_and_discrimination",
        violenceAndThreats: "violence_and_threats",
        dangerousAndCriminalContent: "dangerous_and_criminal_content",
    });
});
export function moderationLlmv1CategoryThresholdsToJSON(moderationLlmv1CategoryThresholds) {
    return JSON.stringify(ModerationLlmv1CategoryThresholds$outboundSchema.parse(moderationLlmv1CategoryThresholds));
}
export function moderationLlmv1CategoryThresholdsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ModerationLlmv1CategoryThresholds$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ModerationLlmv1CategoryThresholds' from JSON`);
}
