// @mistralai/mistralai/esm/models/components/ocrimageobject.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 851c92a59244
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const OCRImageObject$inboundSchema = z.object({
    id: z.string(),
    top_left_x: z.nullable(z.int()),
    top_left_y: z.nullable(z.int()),
    bottom_right_x: z.nullable(z.int()),
    bottom_right_y: z.nullable(z.int()),
    image_base64: z.nullable(z.string()).optional(),
    image_annotation: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "top_left_x": "topLeftX",
        "top_left_y": "topLeftY",
        "bottom_right_x": "bottomRightX",
        "bottom_right_y": "bottomRightY",
        "image_base64": "imageBase64",
        "image_annotation": "imageAnnotation",
    });
});
export function ocrImageObjectFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OCRImageObject$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OCRImageObject' from JSON`);
}
