// @mistralai/mistralai/esm/models/components/ocrtableobject.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: afd50ac93bee
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { OCRConfidenceScore$inboundSchema, } from "./ocrconfidencescore.js";
/**
 * Format of the table
 */
export const Format = {
    Markdown: "markdown",
    Html: "html",
};
/** @internal */
export const Format$inboundSchema = openEnums
    .inboundSchema(Format);
/** @internal */
export const OCRTableObject$inboundSchema = z.object({
    id: z.string(),
    content: z.string(),
    format: Format$inboundSchema,
    word_confidence_scores: z.nullable(z.array(OCRConfidenceScore$inboundSchema)).optional(),
}).transform((v) => {
    return remap$(v, {
        "word_confidence_scores": "wordConfidenceScores",
    });
});
export function ocrTableObjectFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OCRTableObject$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OCRTableObject' from JSON`);
}
