// @mistralai/mistralai/esm/models/components/queryinvocationbody.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7b6b2574d1bd
 */
import * as z from "zod/v4";
import { smartUnion } from "../../types/smartUnion.js";
import { NetworkEncodedInput$outboundSchema, } from "./networkencodedinput.js";
/** @internal */
export const QueryInvocationBodyInput$outboundSchema = smartUnion([
    NetworkEncodedInput$outboundSchema,
    z.record(z.string(), z.any()),
]);
export function queryInvocationBodyInputToJSON(queryInvocationBodyInput) {
    return JSON.stringify(QueryInvocationBodyInput$outboundSchema.parse(queryInvocationBodyInput));
}
/** @internal */
export const QueryInvocationBody$outboundSchema = z.object({
    name: z.string(),
    input: z.nullable(smartUnion([
        NetworkEncodedInput$outboundSchema,
        z.record(z.string(), z.any()),
    ])).optional(),
});
export function queryInvocationBodyToJSON(queryInvocationBody) {
    return JSON.stringify(QueryInvocationBody$outboundSchema.parse(queryInvocationBody));
}
