// @mistralai/mistralai/esm/models/components/realtimetranscriptionsessionupdated.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 50019fc8a613
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { RealtimeTranscriptionSession$inboundSchema, RealtimeTranscriptionSession$outboundSchema, } from "./realtimetranscriptionsession.js";
/** @internal */
export const RealtimeTranscriptionSessionUpdated$inboundSchema = z.object({
    type: z.literal("session.updated").default("session.updated"),
    session: RealtimeTranscriptionSession$inboundSchema,
});
/** @internal */
export const RealtimeTranscriptionSessionUpdated$outboundSchema = z.object({
    type: z.literal("session.updated").default("session.updated"),
    session: RealtimeTranscriptionSession$outboundSchema,
});
export function realtimeTranscriptionSessionUpdatedToJSON(realtimeTranscriptionSessionUpdated) {
    return JSON.stringify(RealtimeTranscriptionSessionUpdated$outboundSchema.parse(realtimeTranscriptionSessionUpdated));
}
export function realtimeTranscriptionSessionUpdatedFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionSessionUpdated$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionSessionUpdated' from JSON`);
}
