// @mistralai/mistralai/esm/models/components/realtimetranscriptionsession.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 33c48d7d6031
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { AudioFormat$inboundSchema, AudioFormat$outboundSchema, } from "./audioformat.js";
/** @internal */
export const RealtimeTranscriptionSession$inboundSchema = z.object({
    request_id: z.string(),
    model: z.string(),
    audio_format: AudioFormat$inboundSchema,
    target_streaming_delay_ms: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        "request_id": "requestId",
        "audio_format": "audioFormat",
        "target_streaming_delay_ms": "targetStreamingDelayMs",
    });
});
/** @internal */
export const RealtimeTranscriptionSession$outboundSchema = z.object({
    requestId: z.string(),
    model: z.string(),
    audioFormat: AudioFormat$outboundSchema,
    targetStreamingDelayMs: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        requestId: "request_id",
        audioFormat: "audio_format",
        targetStreamingDelayMs: "target_streaming_delay_ms",
    });
});
export function realtimeTranscriptionSessionToJSON(realtimeTranscriptionSession) {
    return JSON.stringify(RealtimeTranscriptionSession$outboundSchema.parse(realtimeTranscriptionSession));
}
export function realtimeTranscriptionSessionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionSession$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionSession' from JSON`);
}
