// @mistralai/mistralai/esm/models/components/realtimetranscriptioninputaudioappend.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: cdca8d10df21
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const RealtimeTranscriptionInputAudioAppend$inboundSchema = z.object({
    type: z.literal("input_audio.append").default("input_audio.append"),
    audio: z.string(),
});
/** @internal */
export const RealtimeTranscriptionInputAudioAppend$outboundSchema = z.object({
    type: z.literal("input_audio.append").default("input_audio.append"),
    audio: z.string(),
});
export function realtimeTranscriptionInputAudioAppendToJSON(realtimeTranscriptionInputAudioAppend) {
    return JSON.stringify(RealtimeTranscriptionInputAudioAppend$outboundSchema.parse(realtimeTranscriptionInputAudioAppend));
}
export function realtimeTranscriptionInputAudioAppendFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionInputAudioAppend$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionInputAudioAppend' from JSON`);
}
