// @mistralai/mistralai/esm/models/components/realtimetranscriptionsessionupdatepayload.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c752122222b7
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { AudioFormat$inboundSchema, AudioFormat$outboundSchema, } from "./audioformat.js";
/** @internal */
export const RealtimeTranscriptionSessionUpdatePayload$inboundSchema = z.object({
    audio_format: z.nullable(AudioFormat$inboundSchema).optional(),
    target_streaming_delay_ms: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        "audio_format": "audioFormat",
        "target_streaming_delay_ms": "targetStreamingDelayMs",
    });
});
/** @internal */
export const RealtimeTranscriptionSessionUpdatePayload$outboundSchema = z.object({
    audioFormat: z.nullable(AudioFormat$outboundSchema).optional(),
    targetStreamingDelayMs: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        audioFormat: "audio_format",
        targetStreamingDelayMs: "target_streaming_delay_ms",
    });
});
export function realtimeTranscriptionSessionUpdatePayloadToJSON(realtimeTranscriptionSessionUpdatePayload) {
    return JSON.stringify(RealtimeTranscriptionSessionUpdatePayload$outboundSchema.parse(realtimeTranscriptionSessionUpdatePayload));
}
export function realtimeTranscriptionSessionUpdatePayloadFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionSessionUpdatePayload$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionSessionUpdatePayload' from JSON`);
}
