// @mistralai/mistralai/esm/models/components/realtimetranscriptionsessioncreated.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 08cf23186690
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { RealtimeTranscriptionSession$inboundSchema, RealtimeTranscriptionSession$outboundSchema, } from "./realtimetranscriptionsession.js";
/** @internal */
export const RealtimeTranscriptionSessionCreated$inboundSchema = z.object({
    type: z.literal("session.created").default("session.created"),
    session: RealtimeTranscriptionSession$inboundSchema,
});
/** @internal */
export const RealtimeTranscriptionSessionCreated$outboundSchema = z.object({
    type: z.literal("session.created").default("session.created"),
    session: RealtimeTranscriptionSession$outboundSchema,
});
export function realtimeTranscriptionSessionCreatedToJSON(realtimeTranscriptionSessionCreated) {
    return JSON.stringify(RealtimeTranscriptionSessionCreated$outboundSchema.parse(realtimeTranscriptionSessionCreated));
}
export function realtimeTranscriptionSessionCreatedFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionSessionCreated$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionSessionCreated' from JSON`);
}
