// @mistralai/mistralai/esm/models/components/realtimetranscriptionsessionupdatemessage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ebfd533c189c
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { RealtimeTranscriptionSessionUpdatePayload$inboundSchema, RealtimeTranscriptionSessionUpdatePayload$outboundSchema, } from "./realtimetranscriptionsessionupdatepayload.js";
/** @internal */
export const RealtimeTranscriptionSessionUpdateMessage$inboundSchema = z.object({
    type: z.literal("session.update").default("session.update"),
    session: RealtimeTranscriptionSessionUpdatePayload$inboundSchema,
});
/** @internal */
export const RealtimeTranscriptionSessionUpdateMessage$outboundSchema = z.object({
    type: z.literal("session.update").default("session.update"),
    session: RealtimeTranscriptionSessionUpdatePayload$outboundSchema,
});
export function realtimeTranscriptionSessionUpdateMessageToJSON(realtimeTranscriptionSessionUpdateMessage) {
    return JSON.stringify(RealtimeTranscriptionSessionUpdateMessage$outboundSchema.parse(realtimeTranscriptionSessionUpdateMessage));
}
export function realtimeTranscriptionSessionUpdateMessageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RealtimeTranscriptionSessionUpdateMessage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RealtimeTranscriptionSessionUpdateMessage' from JSON`);
}
