// @mistralai/mistralai/esm/models/components/servercapabilities.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f9901681c983
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { PromptsCapability$inboundSchema, } from "./promptscapability.js";
import { ResourcesCapability$inboundSchema, } from "./resourcescapability.js";
import { ServerTasksCapability$inboundSchema, } from "./servertaskscapability.js";
import { ToolsCapability$inboundSchema, } from "./toolscapability.js";
/** @internal */
export const ServerCapabilities$inboundSchema = z.object({
    experimental: z.nullable(z.record(z.string(), z.record(z.string(), z.any())))
        .optional(),
    logging: z.nullable(z.record(z.string(), z.any())).optional(),
    prompts: z.nullable(PromptsCapability$inboundSchema).optional(),
    resources: z.nullable(ResourcesCapability$inboundSchema).optional(),
    tools: z.nullable(ToolsCapability$inboundSchema).optional(),
    completions: z.nullable(z.record(z.string(), z.any())).optional(),
    tasks: z.nullable(ServerTasksCapability$inboundSchema).optional(),
}).catchall(z.any());
export function serverCapabilitiesFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ServerCapabilities$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ServerCapabilities' from JSON`);
}
