// @mistralai/mistralai/esm/models/components/speechrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5d434727e189
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { SpeechOutputFormat$outboundSchema, } from "./speechoutputformat.js";
/** @internal */
export const SpeechRequest$outboundSchema = z.object({
    model: z.nullable(z.string()).optional(),
    metadata: z.nullable(z.record(z.string(), z.any())).optional(),
    stream: z.boolean().default(false),
    voiceId: z.nullable(z.string()).optional(),
    refAudio: z.nullable(z.string()).optional(),
    input: z.string(),
    responseFormat: SpeechOutputFormat$outboundSchema.optional(),
}).catchall(z.any()).transform((v) => {
    return {
        ...remap$(v, {
            voiceId: "voice_id",
            refAudio: "ref_audio",
            responseFormat: "response_format",
        }),
    };
});
export function speechRequestToJSON(speechRequest) {
    return JSON.stringify(SpeechRequest$outboundSchema.parse(speechRequest));
}
