// @mistralai/mistralai/esm/models/components/tempotracespan.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5c859e8e4cdc
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { TempoTraceAttribute$inboundSchema, } from "./tempotraceattribute.js";
import { TempoTraceEvent$inboundSchema, } from "./tempotraceevent.js";
import { TempoTraceScopeKind$inboundSchema, } from "./tempotracescopekind.js";
/** @internal */
export const TempoTraceSpan$inboundSchema = z.object({
    traceId: z.string(),
    spanId: z.string(),
    parentSpanId: z.nullable(z.string()).optional(),
    name: z.string(),
    kind: TempoTraceScopeKind$inboundSchema,
    startTimeUnixNano: z.string(),
    endTimeUnixNano: z.string(),
    attributes: z.array(TempoTraceAttribute$inboundSchema).optional(),
    events: z.array(TempoTraceEvent$inboundSchema).optional(),
});
export function tempoTraceSpanFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TempoTraceSpan$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TempoTraceSpan' from JSON`);
}
