// @mistralai/mistralai/esm/models/components/timeseriesmetric.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: aa51183d5878
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
/** @internal */
export const TimeSeriesMetricValue1$inboundSchema = smartUnion([z.int(), z.number()]);
export function timeSeriesMetricValue1FromJSON(jsonString) {
    return safeParse(jsonString, (x) => TimeSeriesMetricValue1$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TimeSeriesMetricValue1' from JSON`);
}
/** @internal */
export const TimeSeriesMetricValue2$inboundSchema = smartUnion([z.int(), smartUnion([z.int(), z.number()])]);
export function timeSeriesMetricValue2FromJSON(jsonString) {
    return safeParse(jsonString, (x) => TimeSeriesMetricValue2$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TimeSeriesMetricValue2' from JSON`);
}
/** @internal */
export const TimeSeriesMetric$inboundSchema = z.object({
    value: z.array(z.array(smartUnion([z.int(), smartUnion([z.int(), z.number()])]))),
});
export function timeSeriesMetricFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TimeSeriesMetric$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TimeSeriesMetric' from JSON`);
}
