// @mistralai/mistralai/esm/models/components/toolconfiguration.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7af086bd4464
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const ToolConfiguration$inboundSchema = z.object({
    exclude: z.nullable(z.array(z.string())).optional(),
    include: z.nullable(z.array(z.string())).optional(),
    requires_confirmation: z.nullable(z.array(z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        "requires_confirmation": "requiresConfirmation",
    });
});
/** @internal */
export const ToolConfiguration$outboundSchema = z.object({
    exclude: z.nullable(z.array(z.string())).optional(),
    include: z.nullable(z.array(z.string())).optional(),
    requiresConfirmation: z.nullable(z.array(z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        requiresConfirmation: "requires_confirmation",
    });
});
export function toolConfigurationToJSON(toolConfiguration) {
    return JSON.stringify(ToolConfiguration$outboundSchema.parse(toolConfiguration));
}
export function toolConfigurationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ToolConfiguration$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ToolConfiguration' from JSON`);
}
