// @mistralai/mistralai/esm/models/components/toolexecutiondoneevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9cd620f71513
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import { BuiltInConnectors$inboundSchema, } from "./builtinconnectors.js";
/** @internal */
export const ToolExecutionDoneEventName$inboundSchema = smartUnion([BuiltInConnectors$inboundSchema, z.string()]);
export function toolExecutionDoneEventNameFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ToolExecutionDoneEventName$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ToolExecutionDoneEventName' from JSON`);
}
/** @internal */
export const ToolExecutionDoneEvent$inboundSchema = z.object({
    type: z.literal("tool.execution.done"),
    created_at: z.iso.datetime({ offset: true }).transform(v => new Date(v))
        .optional(),
    output_index: z.int().default(0),
    id: z.string(),
    name: smartUnion([BuiltInConnectors$inboundSchema, z.string()]),
    info: z.record(z.string(), z.any()).optional(),
}).transform((v) => {
    return remap$(v, {
        "created_at": "createdAt",
        "output_index": "outputIndex",
    });
});
export function toolExecutionDoneEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ToolExecutionDoneEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ToolExecutionDoneEvent' from JSON`);
}
