// @mistralai/mistralai/esm/models/components/transcriptionstreamlanguage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ab52ad6e0ffc
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const TranscriptionStreamLanguage$inboundSchema = z.object({
    type: z.literal("transcription.language"),
    audio_language: z.string(),
}).catchall(z.any()).transform((v) => {
    return remap$(v, {
        "audio_language": "audioLanguage",
    });
});
/** @internal */
export const TranscriptionStreamLanguage$outboundSchema = z.object({
    type: z.literal("transcription.language"),
    audioLanguage: z.string(),
}).catchall(z.any()).transform((v) => {
    return {
        ...remap$(v, {
            audioLanguage: "audio_language",
        }),
    };
});
export function transcriptionStreamLanguageToJSON(transcriptionStreamLanguage) {
    return JSON.stringify(TranscriptionStreamLanguage$outboundSchema.parse(transcriptionStreamLanguage));
}
export function transcriptionStreamLanguageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TranscriptionStreamLanguage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TranscriptionStreamLanguage' from JSON`);
}
