// @mistralai/mistralai/esm/models/components/updateconnectorrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: acb3cd06c21b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { AuthData$outboundSchema, } from "./authdata.js";
import { ConnectorProtocol$outboundSchema, } from "./connectorprotocol.js";
/** @internal */
export const UpdateConnectorRequest$outboundSchema = z.object({
    title: z.nullable(z.string()).optional(),
    name: z.nullable(z.string()).optional(),
    description: z.nullable(z.string()).optional(),
    iconUrl: z.nullable(z.string()).optional(),
    systemPrompt: z.nullable(z.string()).optional(),
    protocol: z.nullable(ConnectorProtocol$outboundSchema).optional(),
    connectionConfig: z.nullable(z.record(z.string(), z.any())).optional(),
    connectionSecrets: z.nullable(z.record(z.string(), z.any())).optional(),
    server: z.nullable(z.string()).optional(),
    headers: z.nullable(z.record(z.string(), z.any())).optional(),
    authData: z.nullable(AuthData$outboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        iconUrl: "icon_url",
        systemPrompt: "system_prompt",
        connectionConfig: "connection_config",
        connectionSecrets: "connection_secrets",
        authData: "auth_data",
    });
});
export function updateConnectorRequestToJSON(updateConnectorRequest) {
    return JSON.stringify(UpdateConnectorRequest$outboundSchema.parse(updateConnectorRequest));
}
