// @mistralai/mistralai/esm/models/components/updateinvocationbody.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a646f642f382
 */
import * as z from "zod/v4";
import { smartUnion } from "../../types/smartUnion.js";
import { NetworkEncodedInput$outboundSchema, } from "./networkencodedinput.js";
/** @internal */
export const UpdateInvocationBodyInput$outboundSchema = smartUnion([
    NetworkEncodedInput$outboundSchema,
    z.record(z.string(), z.any()),
]);
export function updateInvocationBodyInputToJSON(updateInvocationBodyInput) {
    return JSON.stringify(UpdateInvocationBodyInput$outboundSchema.parse(updateInvocationBodyInput));
}
/** @internal */
export const UpdateInvocationBody$outboundSchema = z.object({
    name: z.string(),
    input: z.nullable(smartUnion([
        NetworkEncodedInput$outboundSchema,
        z.record(z.string(), z.any()),
    ])).optional(),
});
export function updateInvocationBodyToJSON(updateInvocationBody) {
    return JSON.stringify(UpdateInvocationBody$outboundSchema.parse(updateInvocationBody));
}
