// @mistralai/mistralai/esm/models/components/updatejudgerequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 952724672a76
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { JudgeClassificationOutput$outboundSchema, } from "./judgeclassificationoutput.js";
import { JudgeRegressionOutput$outboundSchema, } from "./judgeregressionoutput.js";
/** @internal */
export const UpdateJudgeRequestOutput$outboundSchema = z.union([
    JudgeClassificationOutput$outboundSchema,
    JudgeRegressionOutput$outboundSchema,
]);
export function updateJudgeRequestOutputToJSON(updateJudgeRequestOutput) {
    return JSON.stringify(UpdateJudgeRequestOutput$outboundSchema.parse(updateJudgeRequestOutput));
}
/** @internal */
export const UpdateJudgeRequest$outboundSchema = z.object({
    name: z.string(),
    description: z.string(),
    modelName: z.string(),
    output: z.union([
        JudgeClassificationOutput$outboundSchema,
        JudgeRegressionOutput$outboundSchema,
    ]),
    instructions: z.string(),
    tools: z.array(z.string()),
}).transform((v) => {
    return remap$(v, {
        modelName: "model_name",
    });
});
export function updateJudgeRequestToJSON(updateJudgeRequest) {
    return JSON.stringify(UpdateJudgeRequest$outboundSchema.parse(updateJudgeRequest));
}
