// @mistralai/mistralai/esm/models/components/validationerror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7f5955a97454
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
/** @internal */
export const Loc$inboundSchema = smartUnion([
    z.string(),
    z.int(),
]);
export function locFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Loc$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Loc' from JSON`);
}
/** @internal */
export const Context$inboundSchema = z.object({});
export function contextFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Context$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Context' from JSON`);
}
/** @internal */
export const ValidationError$inboundSchema = z.object({
    loc: z.array(smartUnion([z.string(), z.int()])),
    msg: z.string(),
    type: z.string(),
    input: z.any().optional(),
    ctx: z.lazy(() => Context$inboundSchema).optional(),
});
export function validationErrorFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ValidationError$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ValidationError' from JSON`);
}
