// @mistralai/mistralai/esm/models/components/vespasearchindexinforesponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f5eab6d3fc97
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { VespaSchemaResponse$inboundSchema, } from "./vespaschemaresponse.js";
/** @internal */
export const VespaSearchIndexInfoResponse$inboundSchema = z.object({
    type: z.literal("vespa"),
    k8s_cluster: z.string(),
    k8s_namespace: z.string(),
    vespa_instance_name: z.string(),
    schemas: z.array(VespaSchemaResponse$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "k8s_cluster": "k8sCluster",
        "k8s_namespace": "k8sNamespace",
        "vespa_instance_name": "vespaInstanceName",
    });
});
export function vespaSearchIndexInfoResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => VespaSearchIndexInfoResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'VespaSearchIndexInfoResponse' from JSON`);
}
