// @mistralai/mistralai/esm/models/components/voicecreaterequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7b19e0a41e95
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { VoiceAppearance$outboundSchema, } from "./voiceappearance.js";
/** @internal */
export const VoiceCreateRequest$outboundSchema = z.object({
    name: z.string(),
    slug: z.nullable(z.string()).optional(),
    languages: z.array(z.string()).optional(),
    gender: z.nullable(z.string()).optional(),
    age: z.nullable(z.int()).optional(),
    tags: z.nullable(z.array(z.string())).optional(),
    color: z.nullable(z.string()).optional(),
    description: z.nullable(z.string()).optional(),
    appearance: z.nullable(VoiceAppearance$outboundSchema).optional(),
    retentionNotice: z.int().default(30),
    sampleAudio: z.string(),
    sampleFilename: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        retentionNotice: "retention_notice",
        sampleAudio: "sample_audio",
        sampleFilename: "sample_filename",
    });
});
export function voiceCreateRequestToJSON(voiceCreateRequest) {
    return JSON.stringify(VoiceCreateRequest$outboundSchema.parse(voiceCreateRequest));
}
