// @mistralai/mistralai/esm/models/components/voiceappearance.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ce200ca4f463
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const VoiceAppearance$inboundSchema = z.object({
    skin_color: z.nullable(z.string()).optional(),
    hair_color: z.nullable(z.string()).optional(),
    background_color: z.nullable(z.string()).optional(),
    avatar_id: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "skin_color": "skinColor",
        "hair_color": "hairColor",
        "background_color": "backgroundColor",
        "avatar_id": "avatarId",
    });
});
/** @internal */
export const VoiceAppearance$outboundSchema = z.object({
    skinColor: z.nullable(z.string()).optional(),
    hairColor: z.nullable(z.string()).optional(),
    backgroundColor: z.nullable(z.string()).optional(),
    avatarId: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        skinColor: "skin_color",
        hairColor: "hair_color",
        backgroundColor: "background_color",
        avatarId: "avatar_id",
    });
});
export function voiceAppearanceToJSON(voiceAppearance) {
    return JSON.stringify(VoiceAppearance$outboundSchema.parse(voiceAppearance));
}
export function voiceAppearanceFromJSON(jsonString) {
    return safeParse(jsonString, (x) => VoiceAppearance$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'VoiceAppearance' from JSON`);
}
