// @mistralai/mistralai/esm/models/components/websearchtool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3b81da54d1b2
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ToolConfiguration$inboundSchema, ToolConfiguration$outboundSchema, } from "./toolconfiguration.js";
/** @internal */
export const WebSearchTool$inboundSchema = z
    .object({
    tool_configuration: z.nullable(ToolConfiguration$inboundSchema).optional(),
    type: z.literal("web_search"),
}).transform((v) => {
    return remap$(v, {
        "tool_configuration": "toolConfiguration",
    });
});
/** @internal */
export const WebSearchTool$outboundSchema = z.object({
    toolConfiguration: z.nullable(ToolConfiguration$outboundSchema).optional(),
    type: z.literal("web_search"),
}).transform((v) => {
    return remap$(v, {
        toolConfiguration: "tool_configuration",
    });
});
export function webSearchToolToJSON(webSearchTool) {
    return JSON.stringify(WebSearchTool$outboundSchema.parse(webSearchTool));
}
export function webSearchToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WebSearchTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WebSearchTool' from JSON`);
}
