// @mistralai/mistralai/esm/models/components/websearchpremiumtool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2083e07cf420
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ToolConfiguration$inboundSchema, ToolConfiguration$outboundSchema, } from "./toolconfiguration.js";
/** @internal */
export const WebSearchPremiumTool$inboundSchema = z.object({
    tool_configuration: z.nullable(ToolConfiguration$inboundSchema).optional(),
    type: z.literal("web_search_premium"),
}).transform((v) => {
    return remap$(v, {
        "tool_configuration": "toolConfiguration",
    });
});
/** @internal */
export const WebSearchPremiumTool$outboundSchema = z.object({
    toolConfiguration: z.nullable(ToolConfiguration$outboundSchema).optional(),
    type: z.literal("web_search_premium"),
}).transform((v) => {
    return remap$(v, {
        toolConfiguration: "tool_configuration",
    });
});
export function webSearchPremiumToolToJSON(webSearchPremiumTool) {
    return JSON.stringify(WebSearchPremiumTool$outboundSchema.parse(webSearchPremiumTool));
}
export function webSearchPremiumToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WebSearchPremiumTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WebSearchPremiumTool' from JSON`);
}
