// @mistralai/mistralai/esm/models/components/workflowexecutionstartedattributesresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 22faf5428a28
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { JSONPayloadResponse$inboundSchema, } from "./jsonpayloadresponse.js";
/** @internal */
export const WorkflowExecutionStartedAttributesResponse$inboundSchema = z.object({
    task_id: z.string(),
    workflow_name: z.string(),
    display_name: z.nullable(z.string()).optional(),
    input: JSONPayloadResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "task_id": "taskId",
        "workflow_name": "workflowName",
        "display_name": "displayName",
    });
});
export function workflowExecutionStartedAttributesResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionStartedAttributesResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionStartedAttributesResponse' from JSON`);
}
