// @mistralai/mistralai/esm/models/components/workflowschedulerequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1268562a8b7c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { ScheduleDefinition$outboundSchema, } from "./scheduledefinition.js";
/** @internal */
export const WorkflowScheduleRequest$outboundSchema = z.object({
    schedule: ScheduleDefinition$outboundSchema,
    workflowRegistrationId: z.nullable(z.string()).optional(),
    workflowVersionId: z.nullable(z.string()).optional(),
    workflowIdentifier: z.nullable(z.string()).optional(),
    workflowTaskQueue: z.nullable(z.string()).optional(),
    scheduleId: z.nullable(z.string()).optional(),
    deploymentName: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        workflowRegistrationId: "workflow_registration_id",
        workflowVersionId: "workflow_version_id",
        workflowIdentifier: "workflow_identifier",
        workflowTaskQueue: "workflow_task_queue",
        scheduleId: "schedule_id",
        deploymentName: "deployment_name",
    });
});
export function workflowScheduleRequestToJSON(workflowScheduleRequest) {
    return JSON.stringify(WorkflowScheduleRequest$outboundSchema.parse(workflowScheduleRequest));
}
