// @mistralai/mistralai/esm/models/components/workflowexecutioncontinuedasnewattributesresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 65d6703812ce
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { JSONPayloadResponse$inboundSchema, } from "./jsonpayloadresponse.js";
/** @internal */
export const WorkflowExecutionContinuedAsNewAttributesResponse$inboundSchema = z
    .object({
    task_id: z.string(),
    new_execution_run_id: z.string(),
    workflow_name: z.string(),
    input: JSONPayloadResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "task_id": "taskId",
        "new_execution_run_id": "newExecutionRunId",
        "workflow_name": "workflowName",
    });
});
export function workflowExecutionContinuedAsNewAttributesResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionContinuedAsNewAttributesResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionContinuedAsNewAttributesResponse' from JSON`);
}
