// @mistralai/mistralai/esm/models/components/workflowexecutiontraceeventsresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a08172baad8c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { smartUnion } from "../../types/smartUnion.js";
import { WorkflowExecutionProgressTraceEvent$inboundSchema, } from "./workflowexecutionprogresstraceevent.js";
import { WorkflowExecutionStatus$inboundSchema, } from "./workflowexecutionstatus.js";
import { WorkflowExecutionTraceEvent$inboundSchema, } from "./workflowexecutiontraceevent.js";
/** @internal */
export const WorkflowExecutionTraceEventsResponseEvent$inboundSchema = smartUnion([
    WorkflowExecutionProgressTraceEvent$inboundSchema,
    WorkflowExecutionTraceEvent$inboundSchema,
]);
export function workflowExecutionTraceEventsResponseEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionTraceEventsResponseEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionTraceEventsResponseEvent' from JSON`);
}
/** @internal */
export const WorkflowExecutionTraceEventsResponse$inboundSchema = z.object({
    workflow_name: z.string(),
    execution_id: z.string(),
    parent_execution_id: z.nullable(z.string()).optional(),
    root_execution_id: z.string(),
    run_id: z.nullable(z.string()).optional(),
    status: z.nullable(WorkflowExecutionStatus$inboundSchema),
    start_time: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    end_time: z.nullable(z.iso.datetime({ offset: true }).transform(v => new Date(v))),
    total_duration_ms: z.nullable(z.int()).optional(),
    result: z.nullable(z.any()),
    events: z.array(smartUnion([
        WorkflowExecutionProgressTraceEvent$inboundSchema,
        WorkflowExecutionTraceEvent$inboundSchema,
    ])).optional(),
}).transform((v) => {
    return remap$(v, {
        "workflow_name": "workflowName",
        "execution_id": "executionId",
        "parent_execution_id": "parentExecutionId",
        "root_execution_id": "rootExecutionId",
        "run_id": "runId",
        "start_time": "startTime",
        "end_time": "endTime",
        "total_duration_ms": "totalDurationMs",
    });
});
export function workflowExecutionTraceEventsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionTraceEventsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionTraceEventsResponse' from JSON`);
}
