// @mistralai/mistralai/esm/models/components/workflow.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 76b9ea763f34
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { WorkflowType$inboundSchema } from "./workflowtype.js";
/** @internal */
export const Workflow$inboundSchema = z.object({
    id: z.string(),
    name: z.string(),
    display_name: z.string(),
    type: WorkflowType$inboundSchema,
    description: z.nullable(z.string()).optional(),
    customer_id: z.string(),
    workspace_id: z.string(),
    shared_namespace: z.nullable(z.string()).optional(),
    available_in_chat_assistant: z.boolean().default(false),
    is_technical: z.boolean().default(false),
    on_behalf_of: z.boolean().default(false),
    archived: z.boolean().default(false),
}).transform((v) => {
    return remap$(v, {
        "display_name": "displayName",
        "customer_id": "customerId",
        "workspace_id": "workspaceId",
        "shared_namespace": "sharedNamespace",
        "available_in_chat_assistant": "availableInChatAssistant",
        "is_technical": "isTechnical",
        "on_behalf_of": "onBehalfOf",
    });
});
export function workflowFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Workflow$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Workflow' from JSON`);
}
