// @mistralai/mistralai/esm/models/components/workflowexecutionrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: bb6034131d8e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
/** @internal */
export const WorkflowExecutionRequest$outboundSchema = z.object({
    executionId: z.nullable(z.string()).optional(),
    input: z.nullable(z.any()).optional(),
    waitForResult: z.boolean().default(false),
    timeoutSeconds: z.nullable(z.number()).optional(),
    customTracingAttributes: z.nullable(z.record(z.string(), z.string()))
        .optional(),
    extensions: z.nullable(z.record(z.string(), z.any())).optional(),
    taskQueue: z.nullable(z.string()).optional(),
    deploymentName: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        executionId: "execution_id",
        waitForResult: "wait_for_result",
        timeoutSeconds: "timeout_seconds",
        customTracingAttributes: "custom_tracing_attributes",
        taskQueue: "task_queue",
        deploymentName: "deployment_name",
    });
});
export function workflowExecutionRequestToJSON(workflowExecutionRequest) {
    return JSON.stringify(WorkflowExecutionRequest$outboundSchema.parse(workflowExecutionRequest));
}
