// @mistralai/mistralai/esm/models/components/workflowexecutiontraceevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3aede01e2247
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { EventType$inboundSchema } from "./eventtype.js";
import { WorkflowExecutionTraceSummaryAttributesValues$inboundSchema, } from "./workflowexecutiontracesummaryattributesvalues.js";
/** @internal */
export const WorkflowExecutionTraceEvent$inboundSchema = z.object({
    type: EventType$inboundSchema.optional(),
    name: z.string(),
    id: z.string(),
    timestamp_unix_nano: z.int(),
    attributes: z.record(z.string(), z.nullable(WorkflowExecutionTraceSummaryAttributesValues$inboundSchema)),
    internal: z.boolean().default(false),
}).transform((v) => {
    return remap$(v, {
        "timestamp_unix_nano": "timestampUnixNano",
    });
});
export function workflowExecutionTraceEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionTraceEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionTraceEvent' from JSON`);
}
