// @mistralai/mistralai/esm/models/components/workflowregistrationwithworkerstatus.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2c5b14d0424b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Workflow$inboundSchema } from "./workflow.js";
import { WorkflowCodeDefinition$inboundSchema, } from "./workflowcodedefinition.js";
/** @internal */
export const WorkflowRegistrationWithWorkerStatus$inboundSchema = z.object({
    id: z.string(),
    deployment_id: z.nullable(z.string()).optional(),
    task_queue: z.nullable(z.string()).optional(),
    definition: WorkflowCodeDefinition$inboundSchema,
    workflow_id: z.string(),
    workflow: z.nullable(Workflow$inboundSchema).optional(),
    compatible_with_chat_assistant: z.boolean().default(false),
    active: z.boolean(),
}).transform((v) => {
    return remap$(v, {
        "deployment_id": "deploymentId",
        "task_queue": "taskQueue",
        "workflow_id": "workflowId",
        "compatible_with_chat_assistant": "compatibleWithChatAssistant",
    });
});
export function workflowRegistrationWithWorkerStatusFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowRegistrationWithWorkerStatus$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowRegistrationWithWorkerStatus' from JSON`);
}
