// @mistralai/mistralai/esm/models/components/workflowexecutionfailedresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6dca52023a89
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { WorkflowExecutionFailedAttributes$inboundSchema, } from "./workflowexecutionfailedattributes.js";
/** @internal */
export const WorkflowExecutionFailedResponse$inboundSchema = z.object({
    event_id: z.string(),
    event_timestamp: z.int(),
    root_workflow_exec_id: z.string(),
    parent_workflow_exec_id: z.nullable(z.string()),
    workflow_exec_id: z.string(),
    workflow_run_id: z.string(),
    workflow_name: z.string(),
    event_type: z.literal("WORKFLOW_EXECUTION_FAILED").default("WORKFLOW_EXECUTION_FAILED"),
    attributes: WorkflowExecutionFailedAttributes$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "event_id": "eventId",
        "event_timestamp": "eventTimestamp",
        "root_workflow_exec_id": "rootWorkflowExecId",
        "parent_workflow_exec_id": "parentWorkflowExecId",
        "workflow_exec_id": "workflowExecId",
        "workflow_run_id": "workflowRunId",
        "workflow_name": "workflowName",
        "event_type": "eventType",
    });
});
export function workflowExecutionFailedResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionFailedResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionFailedResponse' from JSON`);
}
