// @mistralai/mistralai/esm/models/components/workflowexecutiontracesummaryresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5e42d1cb2dcb
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { WorkflowExecutionStatus$inboundSchema, } from "./workflowexecutionstatus.js";
import { WorkflowExecutionTraceSummarySpan$inboundSchema, } from "./workflowexecutiontracesummaryspan.js";
/** @internal */
export const WorkflowExecutionTraceSummaryResponse$inboundSchema = z.object({
    workflow_name: z.string(),
    execution_id: z.string(),
    parent_execution_id: z.nullable(z.string()).optional(),
    root_execution_id: z.string(),
    run_id: z.nullable(z.string()).optional(),
    status: z.nullable(WorkflowExecutionStatus$inboundSchema),
    start_time: z.iso.datetime({ offset: true }).transform(v => new Date(v)),
    end_time: z.nullable(z.iso.datetime({ offset: true }).transform(v => new Date(v))),
    total_duration_ms: z.nullable(z.int()).optional(),
    result: z.nullable(z.any()),
    span_tree: z.nullable(WorkflowExecutionTraceSummarySpan$inboundSchema)
        .optional(),
}).transform((v) => {
    return remap$(v, {
        "workflow_name": "workflowName",
        "execution_id": "executionId",
        "parent_execution_id": "parentExecutionId",
        "root_execution_id": "rootExecutionId",
        "run_id": "runId",
        "start_time": "startTime",
        "end_time": "endTime",
        "total_duration_ms": "totalDurationMs",
        "span_tree": "spanTree",
    });
});
export function workflowExecutionTraceSummaryResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionTraceSummaryResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionTraceSummaryResponse' from JSON`);
}
