// @mistralai/mistralai/esm/models/components/workflowexecutioncontinuedasnewresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d87df6f4d662
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { WorkflowExecutionContinuedAsNewAttributesResponse$inboundSchema, } from "./workflowexecutioncontinuedasnewattributesresponse.js";
/** @internal */
export const WorkflowExecutionContinuedAsNewResponse$inboundSchema = z.object({
    event_id: z.string(),
    event_timestamp: z.int(),
    root_workflow_exec_id: z.string(),
    parent_workflow_exec_id: z.nullable(z.string()),
    workflow_exec_id: z.string(),
    workflow_run_id: z.string(),
    workflow_name: z.string(),
    event_type: z.literal("WORKFLOW_EXECUTION_CONTINUED_AS_NEW").default("WORKFLOW_EXECUTION_CONTINUED_AS_NEW"),
    attributes: WorkflowExecutionContinuedAsNewAttributesResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "event_id": "eventId",
        "event_timestamp": "eventTimestamp",
        "root_workflow_exec_id": "rootWorkflowExecId",
        "parent_workflow_exec_id": "parentWorkflowExecId",
        "workflow_exec_id": "workflowExecId",
        "workflow_run_id": "workflowRunId",
        "workflow_name": "workflowName",
        "event_type": "eventType",
    });
});
export function workflowExecutionContinuedAsNewResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionContinuedAsNewResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionContinuedAsNewResponse' from JSON`);
}
