// @mistralai/mistralai/esm/models/components/workflowexecutioncanceledresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a1656c0c629a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { WorkflowExecutionCanceledAttributes$inboundSchema, } from "./workflowexecutioncanceledattributes.js";
/** @internal */
export const WorkflowExecutionCanceledResponse$inboundSchema = z.object({
    event_id: z.string(),
    event_timestamp: z.int(),
    root_workflow_exec_id: z.string(),
    parent_workflow_exec_id: z.nullable(z.string()),
    workflow_exec_id: z.string(),
    workflow_run_id: z.string(),
    workflow_name: z.string(),
    event_type: z.literal("WORKFLOW_EXECUTION_CANCELED").default("WORKFLOW_EXECUTION_CANCELED"),
    attributes: WorkflowExecutionCanceledAttributes$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "event_id": "eventId",
        "event_timestamp": "eventTimestamp",
        "root_workflow_exec_id": "rootWorkflowExecId",
        "parent_workflow_exec_id": "parentWorkflowExecId",
        "workflow_exec_id": "workflowExecId",
        "workflow_run_id": "workflowRunId",
        "workflow_name": "workflowName",
        "event_type": "eventType",
    });
});
export function workflowExecutionCanceledResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WorkflowExecutionCanceledResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WorkflowExecutionCanceledResponse' from JSON`);
}
