// @mistralai/mistralai/src/models/operations/connectordeactivatefororganizationv1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8e86163c41b4
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type ConnectorDeactivateForOrganizationV1Request = {
  connectorId: string;
};

/** @internal */
export type ConnectorDeactivateForOrganizationV1Request$Outbound = {
  connector_id: string;
};

/** @internal */
export const ConnectorDeactivateForOrganizationV1Request$outboundSchema:
  z.ZodType<
    ConnectorDeactivateForOrganizationV1Request$Outbound,
    ConnectorDeactivateForOrganizationV1Request
  > = z.object({
    connectorId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      connectorId: "connector_id",
    });
  });

export function connectorDeactivateForOrganizationV1RequestToJSON(
  connectorDeactivateForOrganizationV1Request:
    ConnectorDeactivateForOrganizationV1Request,
): string {
  return JSON.stringify(
    ConnectorDeactivateForOrganizationV1Request$outboundSchema.parse(
      connectorDeactivateForOrganizationV1Request,
    ),
  );
}
