// @mistralai/mistralai/src/models/operations/connectordeactivateforworkspacev1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fd86eabfe29c
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type ConnectorDeactivateForWorkspaceV1Request = {
  connectorId: string;
};

/** @internal */
export type ConnectorDeactivateForWorkspaceV1Request$Outbound = {
  connector_id: string;
};

/** @internal */
export const ConnectorDeactivateForWorkspaceV1Request$outboundSchema: z.ZodType<
  ConnectorDeactivateForWorkspaceV1Request$Outbound,
  ConnectorDeactivateForWorkspaceV1Request
> = z.object({
  connectorId: z.string(),
}).transform((v) => {
  return remap$(v, {
    connectorId: "connector_id",
  });
});

export function connectorDeactivateForWorkspaceV1RequestToJSON(
  connectorDeactivateForWorkspaceV1Request:
    ConnectorDeactivateForWorkspaceV1Request,
): string {
  return JSON.stringify(
    ConnectorDeactivateForWorkspaceV1Request$outboundSchema.parse(
      connectorDeactivateForWorkspaceV1Request,
    ),
  );
}
