// @mistralai/mistralai/src/models/operations/connectoractivatefororganizationv1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c20e2c702242
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type ConnectorActivateForOrganizationV1Request = {
  connectorId: string;
  toolExecutionConfiguration?:
    | components.ToolExecutionConfiguration
    | null
    | undefined;
};

/** @internal */
export type ConnectorActivateForOrganizationV1Request$Outbound = {
  connector_id: string;
  ToolExecutionConfiguration?:
    | components.ToolExecutionConfiguration$Outbound
    | null
    | undefined;
};

/** @internal */
export const ConnectorActivateForOrganizationV1Request$outboundSchema:
  z.ZodType<
    ConnectorActivateForOrganizationV1Request$Outbound,
    ConnectorActivateForOrganizationV1Request
  > = z.object({
    connectorId: z.string(),
    toolExecutionConfiguration: z.nullable(
      components.ToolExecutionConfiguration$outboundSchema,
    ).optional(),
  }).transform((v) => {
    return remap$(v, {
      connectorId: "connector_id",
      toolExecutionConfiguration: "ToolExecutionConfiguration",
    });
  });

export function connectorActivateForOrganizationV1RequestToJSON(
  connectorActivateForOrganizationV1Request:
    ConnectorActivateForOrganizationV1Request,
): string {
  return JSON.stringify(
    ConnectorActivateForOrganizationV1Request$outboundSchema.parse(
      connectorActivateForOrganizationV1Request,
    ),
  );
}
