// @mistralai/mistralai/src/models/operations/connectoractivateforworkspacev1.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c7bdd7602eb8
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type ConnectorActivateForWorkspaceV1Request = {
  connectorId: string;
  toolExecutionConfiguration?:
    | components.ToolExecutionConfiguration
    | null
    | undefined;
};

/** @internal */
export type ConnectorActivateForWorkspaceV1Request$Outbound = {
  connector_id: string;
  ToolExecutionConfiguration?:
    | components.ToolExecutionConfiguration$Outbound
    | null
    | undefined;
};

/** @internal */
export const ConnectorActivateForWorkspaceV1Request$outboundSchema: z.ZodType<
  ConnectorActivateForWorkspaceV1Request$Outbound,
  ConnectorActivateForWorkspaceV1Request
> = z.object({
  connectorId: z.string(),
  toolExecutionConfiguration: z.nullable(
    components.ToolExecutionConfiguration$outboundSchema,
  ).optional(),
}).transform((v) => {
  return remap$(v, {
    connectorId: "connector_id",
    toolExecutionConfiguration: "ToolExecutionConfiguration",
  });
});

export function connectorActivateForWorkspaceV1RequestToJSON(
  connectorActivateForWorkspaceV1Request:
    ConnectorActivateForWorkspaceV1Request,
): string {
  return JSON.stringify(
    ConnectorActivateForWorkspaceV1Request$outboundSchema.parse(
      connectorActivateForWorkspaceV1Request,
    ),
  );
}
