// @mistralai/mistralai/src/models/operations/getschedulev1workflowsschedulesscheduleidget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c1083833fde3
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";

export type GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest = {
  scheduleId: string;
};

/** @internal */
export type GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest$Outbound = {
  schedule_id: string;
};

/** @internal */
export const GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest$outboundSchema:
  z.ZodType<
    GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest$Outbound,
    GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest
  > = z.object({
    scheduleId: z.string(),
  }).transform((v) => {
    return remap$(v, {
      scheduleId: "schedule_id",
    });
  });

export function getScheduleV1WorkflowsSchedulesScheduleIdGetRequestToJSON(
  getScheduleV1WorkflowsSchedulesScheduleIdGetRequest:
    GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest,
): string {
  return JSON.stringify(
    GetScheduleV1WorkflowsSchedulesScheduleIdGetRequest$outboundSchema.parse(
      getScheduleV1WorkflowsSchedulesScheduleIdGetRequest,
    ),
  );
}
