// @mistralai/mistralai/src/models/operations/resumeschedulev1workflowsschedulesscheduleidresumepost.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: faa2bdaafd32
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest = {
  scheduleId: string;
  workflowSchedulePauseRequest?:
    | components.WorkflowSchedulePauseRequest
    | null
    | undefined;
};

/** @internal */
export type ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest$Outbound =
  {
    schedule_id: string;
    WorkflowSchedulePauseRequest?:
      | components.WorkflowSchedulePauseRequest$Outbound
      | null
      | undefined;
  };

/** @internal */
export const ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest$outboundSchema:
  z.ZodType<
    ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest$Outbound,
    ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest
  > = z.object({
    scheduleId: z.string(),
    workflowSchedulePauseRequest: z.nullable(
      components.WorkflowSchedulePauseRequest$outboundSchema,
    ).optional(),
  }).transform((v) => {
    return remap$(v, {
      scheduleId: "schedule_id",
      workflowSchedulePauseRequest: "WorkflowSchedulePauseRequest",
    });
  });

export function resumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequestToJSON(
  resumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest:
    ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest,
): string {
  return JSON.stringify(
    ResumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest$outboundSchema
      .parse(resumeScheduleV1WorkflowsSchedulesScheduleIdResumePostRequest),
  );
}
