// @mistralai/mistralai/src/models/operations/updateruninfov1ragingestionpipelineconfigurationsidruninfoput.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ec2338d275ae
 */

import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as components from "../components/index.js";

export type UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest =
  {
    id: string;
    updateRunInfo: components.UpdateRunInfo;
  };

/** @internal */
export type UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest$Outbound =
  {
    id: string;
    UpdateRunInfo: components.UpdateRunInfo$Outbound;
  };

/** @internal */
export const UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest$outboundSchema:
  z.ZodType<
    UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest$Outbound,
    UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest
  > = z.object({
    id: z.string(),
    updateRunInfo: components.UpdateRunInfo$outboundSchema,
  }).transform((v) => {
    return remap$(v, {
      updateRunInfo: "UpdateRunInfo",
    });
  });

export function updateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequestToJSON(
  updateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest:
    UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest,
): string {
  return JSON.stringify(
    UpdateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest$outboundSchema
      .parse(
        updateRunInfoV1RagIngestionPipelineConfigurationsIdRunInfoPutRequest,
      ),
  );
}
